// ct-pwa/src/lib/SpecialTag/index.jsx
import React, { PureComponent } from 'react'
import PropTypes from 'prop-types';
import classnames from 'classnames'


class SpecialTag extends PureComponent {
    _handleClickAction = () => {
        this.props.handleClick({ name: this.props.tagName, id: this.props.tagId });
    }
    render() {
        const { state, tagName, className } = this.props;
        const labelClass = classnames({
            'bg-c-blue c-white': state === 'active',
            'bg-blue-grey c-blue': state !== 'active'
        },
            'mr-8',
            'mb-8',
            'flex-inline',
            'br2',
            'lh-solid',
            'box-border',
            'fs-13',
            'pb-10', 'pt-10', 'pr-15', 'pl-15',
            className
        )

        return (
            <p className={ labelClass } onClick={ this._handleClickAction }>
                { tagName }
            </p>
        )
    }
}
SpecialTag.propTypes = {
    className: PropTypes.string,
    type: PropTypes.string,
    handleClick: PropTypes.func,
}
SpecialTag.defaultProps = {
    className: '',
    type: '',
    handleClick: () => { },
    state : 'inActive'
}

export default SpecialTag;
