// ct-pwa/src/components/Trains/IRCTC/index.jsx
import React, { PureComponent } from 'react'
import PropTypes from 'prop-types'
import classNames from 'classnames'

import Header from 'Lib/Header'
import Button from 'Lib/Buttons'
import { Pagetitle } from 'Lib/Subheader'
import { List, ListItem } from 'Lib/List'
import { RadioList, RadioItem } from 'Lib/Radio'
import Sticky from 'Lib/Sticky'

import { openInNewTab, isEmpty } from 'Utils'

import PrePayment from './PrePayment'

class IRCTC extends PureComponent {
  constructor() {
    super()
    this.state = {
      passwordStatus: null
    }
  }

  _handlePasswordStatusChange = status => {
    this.setState({ passwordStatus: status })
  }

  _handleFormSubmission = () => {
    const { passwordStatus } = this.state
    if (!isEmpty(passwordStatus)) {
      if (passwordStatus === 'forgot') {
        openInNewTab('https://irctc.co.in')
      } else {
        this.formRef.submit()
      }
    }
  }

  _handleRef = formRef => {
    this.formRef = formRef
  }

  render() {
    const pageTitle = 'Processing Payment'
    const { passwordStatus } = this.state
    const buttonType = passwordStatus === 'forgot' ? 'secondary' : 'primary'
    const buttonText =
      passwordStatus === 'forgot' ? 'Reset Password' : 'Continue'
    return (
      <div>
        <Sticky>
          <Header menu={true} />
          <Pagetitle title={pageTitle} />
        </Sticky>

        <p className="pl-16 pr-16 mt-20 fw-600 fs-16 lh-title mb-10">
          Updated booking flow as per IRCTC regulations
        </p>
        <PrePayment refHandler={this._handleRef} />
        <List>
          <ListItem>
            <ol className="pl-16 fs-14 pt-10 pb-10">
              <li className="mb-10">Make payment for your booking</li>
              <li className="mb-10 lh-copy">
                Login to IRCTC account with username {this.props.irctcUserName}.
                Your IRCTC login password will be required
              </li>
              <li>Confirm your booking</li>
            </ol>
          </ListItem>
        </List>

        <RadioList
          value={this.state.passwordStatus}
          onChange={this._handlePasswordStatusChange}
          className="mt-30"
        >
          <RadioItem value="remember" className="flex flex-middle h-48">
            {selected => (
              <p className={classNames('fs-15', { 'fw-600': selected })}>
                I remember my IRCTC password
              </p>
            )}
          </RadioItem>
          <RadioItem value="forgot" className="flex flex-middle h-48">
            {selected => (
              <p className={classNames('fs-15', { 'fw-600': selected })}>
                I dont remember my IRCTC password
              </p>
            )}
          </RadioItem>
        </RadioList>

        <div className="mt-20 mb-20 pl-16 pr-16">
          <Button
            disabled={passwordStatus === null}
            onClick={this._handleFormSubmission}
            type={buttonType}
          >
            {buttonText}
          </Button>
        </div>
      </div>
    )
  }
}

IRCTC.propTypes = {
  irctcUserName: PropTypes.string
}

IRCTC.defaultProps = {
  irctcUserName: ''
}

export default IRCTC
