// ct-pwa/src/components/Alert/index.jsx
import React, { PureComponent } from 'react'
import PropTypes from 'prop-types'
import classnames from 'classnames'

import Ripple from 'Lib/Animate/Ripple/'
import Bell from './icons/bell.svg'
import Calendar from './icons/calendar.svg'
import FabTick from './icons/fab-active.svg'

class Alert extends PureComponent {
  render() {
    const {
      className,
      enableRipple,
      onClick,
      fareAlert,
      isUserHavingEmail,
      type
    } = this.props
    const btnClass = classnames('Alert', className)

    let icon
    switch (type) {
      case 'calendar':
        icon = (
          <Calendar
            width="20"
            height="20"
            fill="#339900"
            className="Alert__icon"
          />
        )
        break
      // default Bell icon
      default:
        icon = (
          <Bell 
            width="20" 
            height="20" 
            fill="#339900" 
            className="Alert__icon" 
          />
        )
        break
    }

    return (
      <Ripple enableRipple={enableRipple}>
        {(rippleProps, RippleContent) => {
          return (
            <button className={btnClass} onClick={onClick} {...rippleProps}>
              <Choose>
                <When
                  condition={fareAlert && fareAlert.flag && isUserHavingEmail}
                >
                  <div>
                    {icon}
                    <FabTick
                      width="20"
                      height="20"
                      fill="#339900"
                      className="Alert__icon--active"
                    />
                  </div>
                </When>
                <Otherwise>{icon}</Otherwise>
              </Choose>

              {RippleContent}
            </button>
          )
        }}
      </Ripple>
    )
  }
}

Alert.propTypes = {
  children: PropTypes.string,
  className: PropTypes.string,
  filter: PropTypes.bool,
  onClick: PropTypes.func,
  enableRipple: PropTypes.bool,
  fareAlert: PropTypes.objectOf(PropTypes.any)
}

Alert.defaultProps = {
  children: 'Button',
  filter: false,
  className: '',
  onClick: () => {},
  enableRipple: true,
  fareAlert: {}
}

export default Alert
