// ct-pwa/src/components/Hotels/Results/StarRating/index.jsx
/*eslint-disable*/
import React, { PureComponent } from 'react'
import PropTypes from 'prop-types'
import range from 'lodash.range'
import { List, ListItem } from 'Lib/List'
import StarIcon from './icons/star.svg'

class StarRating extends PureComponent {
  constructor(props) {
    super(props)

    this.state = {
      value: props.value || []
    }

    this.state.config = {
      count: props.count,
      // default color of inactive star
      colorInactive: props.colorInactive,
      classInactive: props.classInactive,
      // color of an active star
      colorActive: props.colorActive,
      classActive: props.classActive
    }
  }

  handleStarClick = event => {
    const currentValue = Number(event.currentTarget.getAttribute('data-index'))
    const { value } = this.state
    const newValue = value.includes(currentValue)
      ? value.filter(v => v !== currentValue)
      : [...value, currentValue]
    this.setState({
      value: newValue
    })
    this.props.onChange(newValue)
  }

  renderStars = () => {
    const { count } = this.props
    const { value, config } = this.state
    const { colorActive, colorInactive, classActive, classInactive } = config
    return range(count).map(i => {
      const selected = value.includes(i + 1)
      return (
        <div
          className="flex flex-column flex-center flex-middle"
          key={i + 1}
          data-index={i + 1}
          onClick={this.handleStarClick}
        >
          <StarIcon
            fill={selected ? colorActive : colorInactive}
            width="24"
            height="24"
          />
          {this.props.showValues && (
            <p className="c-black-90 pt-8">{i + 1}</p>
          )}
        </div>
      )
    })
  }
  render() {
    return (
      <List>
        <ListItem className="h-72 flex flex-between">
          {this.renderStars()}
        </ListItem>
      </List>
    )
  }
}

StarRating.propTypes = {
  value: PropTypes.arrayOf(PropTypes.number),
  count: PropTypes.number.isRequired,
  showValues: PropTypes.bool,
  colorInactive: PropTypes.string,
  classInactive: PropTypes.string,
  colorActive: PropTypes.string,
  classActive: PropTypes.string
}

export default StarRating
