// ct-pwa/src/pages/Trains/Availability/index.jsx
import React from 'react'

import { Summary } from 'Lib/Subheader'
import Header from 'Lib/Header'

import { path, fromQueryString } from 'Utils'

import AvailabilityContainer from 'Containers/Trains/Availability'

const Availability = props => {
  const params = path(['location', 'search'], props).slice(1)
  const data = fromQueryString(params)
  const processedData = {}
  data.adults = parseInt(data.adults, 10)
  data.children = parseInt(data.children, 10)
  data.srMen = parseInt(data.srMen, 10)
  data.srWomen = parseInt(data.srWomen, 10)
  processedData.departOn = parseInt(data.departOn, 10)
  processedData.origin = { code: data.oCode, station: data.oStation }
  processedData.destination = { code: data.dCode, station: data.dStation }
  processedData.travellers = {
    adults: data.adults,
    children: data.children,
    srMen: data.srMen,
    srWomen: data.srWomen
  }
  processedData.tClass = data.tClass
  processedData.train = {
    fare: parseFloat(data.fare, 10),
    distance: parseInt(data.distance, 10),
    name: data.tName,
    number: data.tNumber,
    availableClass: data.aClass
  }

  const pageTitle = `${processedData.train.name} (${
    processedData.train.number
  })`
  const pageSubtitle = `${processedData.origin.station} - ${
    processedData.destination.station
  }`
  return (
    <div>
      <Header menu={true} />
      <Summary title={pageTitle} subtitle={pageSubtitle} />
      <AvailabilityContainer query={processedData} />
    </div>
  )
}

Availability.displayName = 'Trains(AvailabilityPage)'

export default Availability
