// ct-pwa/src/components/GstInfo/index.jsx
import React, { PureComponent } from 'react'
import { browserHistory } from 'Utils/history'
import PropTypes from 'prop-types'

import { List, ListItem } from 'Lib/List'
import Input from 'Lib/Form/Input'
import { Shake } from 'Lib/Animate'
import Button from 'Lib/Buttons'
import { gstValidator, isEmpty } from 'Utils'

import { SEARCH_PAGE } from 'Constants/global-paths'

class GstInfo extends PureComponent {
  constructor(props) {
    super(props)
    this.state = {
      number: props.gst.number || '',
      name: props.gst.name || '',
      address: props.gst.address || '',
      animate: {
        number: false,
        name: false,
        address: false
      },
      validateError: {
        number: false,
        name: false,
        address: false
      },
      gstRegex: props.gstRegex || '',
      isToastVisible: false
    }
  }
  componentWillReceiveProps(nextProps) {
    this.state = {
      number: nextProps.gst.number || '',
      name: nextProps.gst.name || '',
      address: nextProps.gst.address || '',
      animate: {
        number: false,
        name: false,
        address: false
      },
      validateError: {
        number: false,
        name: false,
        address: false
      },
      gstRegex: nextProps.gstRegex || ''
    }
    if (
      this.props.itineraryId === '' &&
      nextProps.gst.error &&
      nextProps.gst.error.length > 0
    ) {
      const searchPagePath = SEARCH_PAGE(this.props.type)
      browserHistory.push({
        pathname: searchPagePath
      })
    }
  }
  _inputValidation = (name, fieldName) => {
    let animate = false
    let error = false
    let result
    if (isEmpty(this.state[name])) {
      error = true
      animate = true
      if (!this.state.isToastVisible) {
        this.setState({ isToastVisible: true })
        this.props.handleValidationError({
          message: `${fieldName} is required`
        })
      }
    } else if (!isEmpty(this.state[name])) {
      if (name === 'number') {
        result = gstValidator(this.state[name], this.state.gstRegex)
        error = !result
        animate = !result
        if (!this.state.isToastVisible && !result) {
          this.setState({ isToastVisible: true })
          this.props.handleValidationError({
            message: 'Please enter valid GST number'
          })
        }
      }
    }
    this.setState(prevState => ({
      validateError: { ...prevState.validateError, [name]: error },
      animate: { ...prevState.animate, [name]: animate }
    }))
    if (name === 'number') {
      return result
    }
  }
  _handleInputChange = (value, name) => {
    this.setState({ [name]: value.trim() })
  }
  _handleInputFocus = (element, name) => {
    if (this.state.validateError[name]) {
      this.setState(prevState => ({
        validateError: {
          ...prevState.validateError,
          [name]: !prevState.validateError[name]
        }
      }))
    }
  }
  _isValidGst = () => {
    const { number, name, address } = this.state
    this.setState({ isToastVisible: false })
    const matched = this._inputValidation('number', 'GST Number')
    this._inputValidation('name', 'Holder name')
    this._inputValidation('address', 'Holder address')
    const { animate } = this.state
    if (!isEmpty(animate)) {
      setTimeout(() => {
        this.setState({
          animate: {
            number: false,
            name: false,
            address: false
          }
        })
      }, 500)
    }
    if (matched && !animate.name && !animate.number && !animate.address) {
      this.setState({
        number,
        name,
        address,
        animate: {},
        validateError: {}
      })
      this.props.setTravellersGst(this.state)
    }
  }
  render() {
    const { number, name, address, validateError, animate } = this.state
    return (
      <form>
        <List className="mt-10">
          <Shake animate={animate.number}>
            <Input
              type="text"
              name="number"
              label="GST No."
              labelClassName="fs-15 c-black-90 flex-fix"
              inputClassName="fs-15 c-black-90"
              wrapper={ListItem}
              wrapperClassName="h-48 flex flex-middle"
              value={number}
              onChange={this._handleInputChange}
              hasError={validateError.number}
              onFocus={this._handleInputFocus}
              maxlength="15"
            />
          </Shake>
          <Shake animate={animate.name}>
            <Input
              type="text"
              name="name"
              label="Holder name"
              labelClassName="fs-15 c-black-90 flex-fix"
              inputClassName="fs-15 c-black-90"
              wrapper={ListItem}
              wrapperClassName="h-48 flex flex-middle"
              value={name}
              onChange={this._handleInputChange}
              hasError={validateError.name}
              onFocus={this._handleInputFocus}
            />
          </Shake>
          <Shake animate={animate.address}>
            <Input
              type="text"
              name="address"
              label="Holder address"
              labelClassName="fs-15 c-black-90 flex-fix"
              inputClassName="fs-15 c-black-90"
              wrapper={ListItem}
              wrapperClassName="h-48 flex flex-middle"
              value={address}
              onChange={this._handleInputChange}
              hasError={validateError.address}
              onFocus={this._handleInputFocus}
            />
          </Shake>
        </List>
        <div className="pl-16 pr-16 mt-20 mb-20 ta-c">
          <Button size="full" type="secondary" onClick={this._isValidGst}>
            Continue booking
          </Button>
        </div>
      </form>
    )
  }
}
GstInfo.propTypes = {
  setTravellersGst: PropTypes.func,
  gst: PropTypes.shape({
    number: PropTypes.string,
    name: PropTypes.string,
    address: PropTypes.string,
    error: PropTypes.string
  }),
  itineraryId: PropTypes.string,
  gstRegex: PropTypes.string,
  type: PropTypes.string,
  handleValidationError: PropTypes.func
}
GstInfo.defaultProps = {
  gst: {
    number: '',
    name: '',
    address: '',
    error: ''
  },
  itineraryId: '',
  setTravellersGst: () => {},
  gstRegex: '',
  type: 'trains',
  handleValidationError: () => {}
}
export default GstInfo
