// ct-pwa/src/components/Hotels/Calendar/index.jsx
import React, { PureComponent } from 'react'
import PropTypes from 'prop-types'
import InputBlock from 'Lib/Form/InputBlock'
import { isEmpty, formatDate } from 'Utils'
import { dateType } from 'Constants/types'

import { Shake } from 'Lib/Animate'

class HotelCalendar extends PureComponent {
  _formatDate = date => {
    if (!isEmpty(date)) {
      return formatDate(date)
    }
    return ''
  }

  render() {
    const { hotelCheckInDate, hotelCheckOutDate } = this.props
    return (
      <Shake animate={this.props.animate}>
        <div className="TouchAndGo__group h-80">
          <InputBlock
            label="Check-in"
            fieldClass="TouchAndGo__tappable--left"
            labelClass="fs-15 mb-10"
            valueClass="fs-18 fw-500 c-black-80"
            value={this._formatDate(hotelCheckInDate)}
            onClick={this.props.handleCheckInDate}
            hasError={this.props.animate}
          />
          <InputBlock
            label="Check-out"
            fieldClass="TouchAndGo__tappable--right"
            labelClass="fs-15 mb-10"
            valueClass="fs-18 fw-500 c-black-80"
            value={this._formatDate(hotelCheckOutDate)}
            onClick={this.props.handleCheckOutDate}
            hasError={this.props.animate}
          />
        </div>
      </Shake>
    )
  }
}

HotelCalendar.propTypes = {
  handleCheckInDate: PropTypes.func,
  handleCheckOutDate: PropTypes.func,
  hotelCheckInDate: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.number,
    dateType
  ]),
  hotelCheckOutDate: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.number,
    dateType
  ]),
  animate: PropTypes.bool
}

HotelCalendar.defaultProps = {
  handleCheckInDate: () => {},
  handleCheckOutDate: () => {},
  hotelCheckInDate: '...',
  hotelCheckOutDate: '...',
  animate: true
}

export default HotelCalendar
