// ct-pwa/src/pages/Hotels/Details/index.jsx
import React, { Component } from 'react'
import { Link } from 'react-router'

import { List, ListItem } from 'Lib/List'
import Button from 'Lib/Buttons'
import ShortlistButton from 'Lib/Buttons/ShortlistButton'
import Deal from 'Lib/Deal'
import Scroller from 'Lib/Scroller'
import Slider from 'Lib/Slider'
import ImageCard from 'Lib/ImageCard'
import Sticky from 'Lib/Sticky'
import HotelsCard from 'Lib/HotelsCard'
import SpecialTag from 'Lib/SpecialTag'
import InfoIcon from 'Lib/InfoIcon'
import Modal from 'Lib/Modal'
import Overlay from 'Lib/Overlay'
import DealsModal from 'Lib/DealsModal'
import TaggedModal from 'Lib/TaggedModal'
import { Truncate } from 'Lib/Text'
import { TripAdvisor } from 'Lib/Rating'
import AddReadMore from 'Lib/AddReadMore'

import SummaryHeader from 'components/Hotels/Results/SummaryHeader'

import { goBack } from 'Utils/history'
import { formatCurrency } from 'Utils'
import { formatMeasurementUnit } from 'Utils/hotels'

import Top from './icons/best.svg'
import StarRating from './icons/star.svg'
import TACircle from './icons/ta.svg'
import TAEmptyCircle from './icons/ta-empty.svg'
import TALogo from './icons/ta-logo.colored.svg'
import Calendar from './icons/calendar.svg'
import Breakfast from './icons/breakfast.svg'
import Taxes from './icons/taxes.svg'
import AC from './icons/ac.svg'
import TV from './icons/tv.svg'
import Wifi from './icons/wifi.svg'
import Airport from './icons/airport.svg'
import Parking from './icons/parking.svg'
import Restaurant from './icons/restaurant.svg'
import BackArrow from './icons/back_arrow.svg'

import HotelImage1 from './images/hotel-1.jpg'
import HotelImage2 from './images/hotel-2.jpg'
import HotelImage3 from './images/hotel-3.jpg'
import HotelImage4 from './images/hotel-4.jpg'
import HotelFacade from './images/hotel-facade.jpg'
import HotelFacade2 from './images/hotel-facade-2.jpg'
import HotelFacade3 from './images/hotel-facade-3.jpg'
import MapImage from './images/map.jpg'

import { calculate } from 'specificity';

class Results extends Component {
	constructor(props) {
		super(props)
		this.state = {
			showingFilterSheet: false,
			showDealModal: false,
			showTaggedModal: false,
		}
	}
	closeDealModal = () => {
		this.setState({ showDealModal: false })
	}
	openTaggedModal = () => {
		this.setState({ showTaggedModal: true })
	}
	closeTaggedModal = () => {
		this.setState({ showTaggedModal: false })
		console.log("inside close taggedModal")
		return false;
	}
	render() {
		const handleBackNavigation = () => {
			goBack(backNavigationParams)
		}
		return (
			<div className="bg-white">
				<SummaryHeader />
				<Scroller>
					<TaggedModal slideUp={this.state.showTaggedModal} closeFn={this.closeTaggedModal}/>
					{this.state.showDealModal ?
						<DealsModal
							dealMessage={["Flat 30% off up to 750 on ICICI Bank Credit & Debit Cards only.Use Coupon CTICICI30",
								"Flat 30% off up to 750 on ICICI Bank Credit & Debit Cards only.Use Coupon CTICICI30"]}
							closeModalFn={this.closeDealModal}
						/>
						: ''
					}
					{/* Gallery */}
					<div className="Gallery p-relative">
						<a className="p-absolute ml-15 mt-30">
							<BackArrow fill="#fff" onClick={handleBackNavigation} />
						</a>
						<div
							className="Gallery__hero Gallery__img"
							style={{ backgroundImage: `url(${HotelImage4})` }}
						/>
						<div className="Gallery__reel">
							<div
								className="Gallery__frame Gallery__img"
								style={{ backgroundImage: `url(${HotelImage2})` }}
							/>
							<div
								className="Gallery__frame Gallery__img"
								style={{ backgroundImage: `url(${HotelImage3})` }}
							/>
							<div
								className="Gallery__frame Gallery__img"
								style={{ backgroundImage: `url(${HotelImage1})` }}
							/>
							<div
								className="Gallery__frame Gallery__img"
								style={{ backgroundImage: `url(${HotelImage4})` }}
							/>
						</div>
					</div>

					{/* Deal */}
					<Deal message="Flat 30% off up to 750 on ICICI Bank Credit & Debit Cards only.Use Coupon CTICICI30" />

					<div className="pt-15">
						<h5 className="fs-18 fw-600 c-black-80 mb-5 pl-15 pr-15">
							Vivanta By Taj Lands End
						</h5>

						{/* Price */}
						<List className="bt-0 mt-8">
							<ListItem className="h-64 flex flex-between flex-middle bb-0">
								<div className="flex flex-column flex-center ta-l">
									<ul className="list--separated__offer-block fw-600 tt-u flex c-bright-red flex-middle">
										<li className="fs-11 pb-5">20% OFF</li>
									</ul>
									<p className="flex flex-middle pb-5">
										<p className="fs-17 fw-600">₹4,530</p>
										<span className="fs-13 c-black-40 pl-5">
											<strike>₹ 16,400</strike>
										</span>
									</p>
									<p className="fs-13 c-black-40 w-100p ">Inclusive of all taxes </p>
								</div>
							</ListItem>

							<ListItem className="h-64 flex flex-middle flex-start bb-0">
								<div className="flex flex-column">
									<div className="flex">
										<StarRating fill="#fec22d" width="15" height="15" />
										<StarRating fill="#fec22d" width="15" height="15" />
										<StarRating fill="#fec22d" width="15" height="15" />
										<StarRating fill="#fec22d" width="15" height="15" />
										<StarRating fill="#fec22d" width="15" height="15" />
									</div>
									<p className="fs-11 c-black-40 mt-5">5 star</p>
								</div>

								<div className="flex flex-column pl-15">
									<div className="flex flex-middle">
										<TALogo width="19" height="13" className="mr-5" />
										<TACircle fill="#589443" width="13" height="13" />
										<TACircle fill="#589443" width="13" height="13" />
										<TACircle fill="#589443" width="13" height="13" />
										<TAEmptyCircle fill="#589443" width="13" height="13" />
										<TAEmptyCircle fill="#589443" width="13" height="13" />
									</div>
									<p className="fs-11 c-black-40 mt-5">233 reviews</p>
								</div>

								<div className="flex flex-middle pl-15">
									<div className="d-ib">
										<Top />
									</div>
									<div className="flex flex-column ml-5">
										<p className="fs-14 c-black-80 fw-600">Top 5%</p>
										<p className="fs-11 c-black-40 mt-3 p-truncate-1 lh-title">booked in Bangalore</p>
									</div>
								</div>
							</ListItem>
						</List>
					</div>

					{/* Highlights */}
					<List className="bt-0 mt-8">
						<ListItem className="pt-15 pb-15 bb-0">
							<h5 className="fs-18 fw-600 c-black-80 mb-10">
								Highlights of the property
							</h5>
							<ol className="list list--ordered mt-15">
								<li className="fs-13 c-black-60 pb-5">
									Dining options include rooftop dining at the Pan-Asian
									restaurant—OKO, and 24-hour outdoor dining at 24x7 Restaurant
								</li>
								<li className="fs-13 c-black-60 pb-5">
									Rejuve—the award-winning spa, and their in-house gym
								</li>
								<li className="fs-13 c-black-60 pb-5">
									The enticing all-year outdoor infinity pool
								</li>
							</ol>
						</ListItem>
					</List>

					{/* Location */}
					<List className="bt-0 mt-8 pb-15">
						<ListItem className="h-128 pt-15 bb-0">
							<h5 className="fs-18 fw-600 c-black-80 mb-5">Location</h5>
							<p className="fs-13 c-black-60">
								Near Golf Course, Kumara Krupa High Grounds
							</p>

							<Link to="/m/v2/hotels/details/map">
								<div className="mr--16 ml--16 mt-10">
									<img src={MapImage} alt="map" height="185" className="w-100p mb--2" />
								</div>
							</Link>
						</ListItem>
						<Link to="/m/v2/hotels/details/map">
							<ListItem className="h-48 flex flex-middle bt bb" arrow={true} style={{
								borderColor: '#eaeaea'
							}}>
								<a className="fs-17 c-blue">View full map</a>
							</ListItem>
						</Link>
					</List>

					{/* Reviews */}
					<List className="bt-0 mt-8 pb-15">
						<ListItem className="h-128 pt-15 pb-15 bb-0">
							<h5 className="fs-18 fw-600 c-black-80 mb-15">
								Reviews from travellers
							</h5>

							<div className="flex flex-top">
								<div className="flex flex-column flex-1 flex-between">
									<p className="fs-40 mb-8 c-green">4.0</p>
									<div className="flex flex-middle mb-8">
										<TALogo
											width="21"
											height="17"
											className="mr-5"
											fill="#333333"
										/>
										<TACircle fill="#589443" width="17" height="17" />
										<TACircle fill="#589443" width="17" height="17" />
										<TACircle fill="#589443" width="17" height="17" />
										<TAEmptyCircle fill="#589443" width="17" height="17" />
										<TAEmptyCircle fill="#589443" width="17" height="17" />
									</div>
									<p className="fs-13 c-black-40">based on 233 reviews</p>
								</div>
								<div className="flex flex-column flex-end">
									<div className="Progress flex flex-middle mb-8">
										<p className="Progress_count fs-11 c-green fw-500">5.0</p>
										<div className="Progress__container mr-10 ml-10" style={{width:'115px'}}>
											<div className="Progress__bar d-ib" />
										</div>
										<p className="Progress_count fs-11 c-black-60 w-16 ta-r t-truncate" style={{
											width: '22px'
										}}>12</p>
									</div>
									<div className="Progress flex flex-middle mb-8">
										<p className="Progress_count fs-11 c-green fw-500">5.0</p>
										<div className="Progress__container mr-10 ml-10 " style={{width:'115px'}}>
											<div className="Progress__bar d-ib" />
										</div>
										<p className="Progress_count fs-11 c-black-60 w-16 ta-r t-truncate" style={{
											width: '22px'
										}}>12</p>
									</div>
									<div className="Progress flex flex-middle mb-8">
										<p className="Progress_count fs-11 c-green fw-500">5.0</p>
										<div className="Progress__container mr-10 ml-10 " style={{width:'115px'}}>
											<div className="Progress__bar d-ib" />
										</div>
										<p className="Progress_count fs-11 c-black-60 w-16 ta-r t-truncate" style={{
											width: '22px'
										}}>90</p>
									</div>
									<div className="Progress flex flex-middle mb-8">
										<p className="Progress_count fs-11 c-green fw-500">5.0</p>
										<div className="Progress__container mr-10 ml-10" style={{width:'115px'}}>
											<div className="Progress__bar d-ib" />
										</div>
										<p className="Progress_count fs-11 c-black-60 w-16 ta-r t-truncate" style={{
											width: '22px'
										}}>22</p>
									</div>
									<div className="Progress flex flex-middle">
										<p className="Progress_count fs-11 c-green fw-500">5.0</p>
										<div className="Progress__container mr-10 ml-10" style={{width:'115px'}}>
											<div className="Progress__bar d-ib" />
										</div>
										<p className="Progress_count fs-11 c-black-60 w-16 ta-r t-truncate" style={{
											width: '22px'
										}}>1</p>
									</div>
								</div>
							</div>
						</ListItem>
						<ListItem className="bb-0">
							<h6 className="fs-17 fw-600 c-black-80 mb-8">
								Recent reviews
							</h6>
						</ListItem>
						<Slider>

							<ListItem className="flex flex-column pb-15 pt-15 bg-light-gray br2 bb-0 h-152 maxh-152">
								<h6 className="fs-15 fw-600 c-black-80 mb-8 p-truncate-1">
									Great place to stay with family
								</h6>
								<div className="flex flex-middle mb-8">
									<div className="d-ib w-64">
										<TACircle fill="#589443" width="12" height="12" />
										<TACircle fill="#589443" width="12" height="12" />
										<TACircle fill="#589443" width="12" height="12" />
										<TAEmptyCircle fill="#589443" width="12" height="12" />
										<TAEmptyCircle fill="#589443" width="12" height="12" />
									</div>
									
									<p className="fs-13 c-black-40 ml-5 p-truncate-1">
										Kunal Khurana on Oct 20, 2017
									</p>
								</div>
								<AddReadMore text="Read More" charLength={165} className="fs-13 c-black-90 lh-copy">
								Firstly the service was excellent. Great front desk and housekeeping team was excellent. However what let the hotel down was the tired and low maintained rooms. I had cracks in the tiles in my room and the shower curtain was black and stained (photo). Some of the lights didn’t work nor did the fridge. Based on this I felt the hotel was too expensive as a room at the nearby Ibis was the same price but much better condition. The hotel had mostly local Indian guests which was represented in the breakfast buffet. Once again I have to make a special mention of reception team who were outstanding. 
								</AddReadMore>
							</ListItem>
							<ListItem className="flex flex-column pb-15 pt-15 bg-light-gray br2 bb-0 h-152 maxh-152">
								<h6 className="fs-15 fw-600 c-black-80 mb-8 p-truncate-1">
									Great place to stay with family
								</h6>
								<div className="flex flex-middle mb-8">
									<div className="d-ib w-64">
										<TACircle fill="#589443" width="12" height="12" />
										<TACircle fill="#589443" width="12" height="12" />
										<TACircle fill="#589443" width="12" height="12" />
										<TAEmptyCircle fill="#589443" width="12" height="12" />
										<TAEmptyCircle fill="#589443" width="12" height="12" />
									</div>
									<p className="fs-13 c-black-40 ml-5 p-truncate-1">
										Kunal Khurana on Oct 20, 2017
									</p>
								</div>
								<AddReadMore text="Read More" charLength={165} className="fs-13 c-black-90 lh-copy">
										I had to stay for almost a month for some official work in
										Bangalore. I chose The Lalit Ashok for its pristine lush green
										surroundings. You can not ask for more as far as the location.
										surroundings You can not ask for more as far as the location
								</AddReadMore>
							</ListItem>
							<ListItem className="flex flex-column pb-15 pt-15 bg-light-gray br2 bb-0 h-152 maxh-152">
								<h6 className="fs-15 fw-600 c-black-80 mb-8 p-truncate-1">
									Great place to stay with family
								</h6>
								<div className="flex flex-middle mb-8">
								<div className="d-ib w-64">
									<TACircle fill="#589443" width="12" height="12" />
									<TACircle fill="#589443" width="12" height="12" />
									<TACircle fill="#589443" width="12" height="12" />
									<TAEmptyCircle fill="#589443" width="12" height="12" />
									<TAEmptyCircle fill="#589443" width="12" height="12" />
								</div>
									
									<p className="fs-13 c-black-40 ml-5 p-truncate-1">
										Kunal Khurana on Oct 20, 2017
									</p>
								</div>
								<AddReadMore charLength={165} className="fs-13 c-black-90 lh-copy">
										I had to stay for almost a month for some official work in
										Bangalore.
								</AddReadMore>
							</ListItem>
						</Slider>
						<Link to="/m/v2/hotels/details/reviews">
							<ListItem className="bb-0" arrow={true}>
								<a className="h-48 flex c-blue flex-middle mt-15 bt fs-17 bb">Read all reviews</a>
							</ListItem>
						</Link>

					</List>
					<List className="bt-0 mt-8 pb-15">
						<ListItem className="h-128 pt-15 bb-0">
							<h5 className="fs-18 fw-600 c-black-80 mb-15">
								Tagged as
							</h5>
							<div >
								<SpecialTag state="active" handleClick={this.openTaggedModal}>
									Good night life
								</SpecialTag>
								<SpecialTag state="active">
									On the beach
								</SpecialTag>
								<SpecialTag>
									Romantic stays
								</SpecialTag>
								<SpecialTag>
									24 Hours room service
								</SpecialTag>
								<SpecialTag>
									Luxury
								</SpecialTag>
								<SpecialTag>
									Infinity pools
								</SpecialTag>
							</div>
						</ListItem>
					</List>

					{/* About */}
					<List className="bt-0 mt-8 pb-15">
						<ListItem className="h-128 pt-15 pb-15 bb-0">
							<h5 className="fs-18 fw-600 c-black-80 mb-5">
								More about the property
</h5>
							<div className="mt-15 mb-15">
								<img src={HotelFacade} alt="map" height="192px" className="w-100p mb--2 br3" />
							</div>
							<p className="fs-13 c-black-60 lh-copy mb-10">
								This grand property, overlooking the sprawling Bangalore Golf
								Club, boasts of its very own tennis court and lush green lawns.
								The spacious rooms, decked with earthy tones, exude subtle
								sophistication. A serene nook, apt for a luxurious business or
								leisure stay!
</p>
							<p className="fs-13 c-black-60 lh-copy">
								The Lalit Ashok, Bangalore is an ideal venue for both leisure
								and business guests. The spacious…
</p>
						</ListItem>
						<ListItem className="h-56 flex flex-middle flex-between bb-0">
							<div className="pr-15 mr-8">
								<p className="fs-16 fw-600 mb-5">12:00</p>
								<p className="fs-13 c-black-40">Check-out</p>
							</div>
							<div className="pr-15 mr-8">
								<p className="fs-16 fw-600 mb-5">12:00</p>
								<p className="fs-13 c-black-40">Check-in</p>
							</div>
							<div className="pr-15 mr-8">
								<p className="fs-16 fw-600 mb-5">184</p>
								<p className="fs-13 c-black-40">Rooms</p>
							</div>
							<div>
								<p className="fs-16 fw-600 mb-5">3</p>
								<p className="fs-13 c-black-40">Floors</p>
							</div>
						</ListItem>
					</List>

					{/* Amenities */}
					<List className="bt-0 mt-8 pb-15">
						<ListItem className="h-56 pt-15 pb-15 bb-0">
							<h5 className="fs-18 fw-600 c-black-80 mb-15">
								Amenities
</h5>
							<div className="flex flex-middle flex-between">
								<div className="flex flex-column flex-middle">
									<AC width="21" height="21" fill="#333333" />
									<p className="fs-13 c-black-40 mt-5">AC</p>
								</div>
								<div className="flex flex-column flex-middle">
									<Restaurant width="21" height="21" fill="#333333" />
									<p className="fs-13 c-black-40 mt-5">Restaurant</p>
								</div>
								<div className="flex flex-column flex-middle">
									<Parking width="21" height="21" fill="#333333" />
									<p className="fs-13 c-black-40 mt-5">Parking</p>
								</div>
								<div className="flex flex-column flex-middle">
									<Airport width="21" height="21" fill="#333333" />
									<p className="fs-13 c-black-40 mt-5">Pickup</p>
								</div>
								<div className="flex flex-column flex-middle">
									<div className="fs-22 c-blue mb-2">+81</div>
									<p className="fs-13 c-blue mt-5">more</p>
								</div>
							</div>
						</ListItem>
					</List>

					{/* Room Rates */}
					<List className="bt-0 mt-8 pb-15">
						<ListItem className="h-56 pt-15 pb-15 bb-0">
							<h5 className="fs-18 fw-600 c-black-80 mb-15">Travel Dates</h5>

							{/* Travel Dates */}
							<div className="flex flex-middle">
								<div className="bg-blue-grey flex-1 flex flex-middle br2 mr-5 h-56">
									<div className="pl-15 pr-15">
										<Calendar fill="#3366cc" />
									</div>
									<div className="flex flex-column flex-1">
										<p className="fs-10 mb-2 tt-u c-faded-blue">Check-in – 12pm</p>
										<p className="fs-13 c-blue fw-600">Mon, 7th Aug</p>
									</div>
								</div>
								<div className="bg-blue-grey flex-1 flex flex-middle br2 ml-5 h-56">
									<div className="pl-15 pr-15">
										<Calendar fill="#3366cc" />
									</div>
									<div className="flex flex-column flex-1">
										<p className="fs-10 mb-2 tt-u c-faded-blue">Check-out – 12pm</p>
										<p className="fs-13 c-blue fw-600">Tue, 7th Aug</p>
									</div>
								</div>
							</div>

							<div className="bb pt-20 mr--16 ml--16"></div>

							{/* Room type */}
							<div className="mt-15">
								<div className="bg-white pl-15 pr-15 pt-15 pb-15 ml--16 mr--16">
									<h3 className="fs-17 fw-600 c-black-80">Deluxe Room</h3>
									<ul className="list list--inline list--separated mt-8">
										<li className="fs-13 c-black-60">King Bed</li>
										<li className="fs-13 c-black-60">336 sq ft</li>
										<li className="fs-13 c-black-60">Water View</li>
									</ul>
									<ul className="list list--inline list--spaced mt-8">
										<li className="fs-13 c-black-60 flex flex-middle">
											<Wifi width="12" height="12" fill="#333333" />
											<span className="pl-5">Wifi</span>
										</li>
										<li className="fs-13 c-black-60 flex flex-middle">
											<TV width="12" height="12" fill="#333333" />
											<span className="pl-5">Television</span>
										</li>
										<li className="fs-13 c-black-60 flex flex-middle">
											<AC width="12" height="12" fill="#333333" />
											<span className="pl-5">Air Condition</span>
										</li>
										<li className="fs-13 c-blue">+13 more</li>
									</ul>

								</div>
								<div className="ml--16 mr--16">
									<Slider centerMode={true}>
										<ImageCard className="mr-16 br3 ml-16" imgUrl={HotelImage1} />
									</Slider>
								</div>

								{/* Room rate */}
								<div className="mt-10">
									<div className="h-56 flex flex-middle flex-between bb" style={{ borderColor: 'rgb(0,0,0,0.1)', borderWidth: '0.5px' }}>
										<div className="flex flex-column">
											<ul className="list--separated tt-u flex c-bright-red" onClick={() => this.setState({ showDealModal: true })}>
												<li className="fs-11 pb-5 fw-500">20% OFF</li>
												<li className="fs-11 pb-5 fw-500">&#10003; 2 deals applied</li>
											</ul>
											<p className="flex flex-middle pb-5">
												<p className="fs-17 fw-600">₹4,530</p>
												<span className="fs-13 c-black-40 pl-5">
													<strike>₹ 16,400</strike>
												</span>
											</p>
											<p className="fs-13 c-black-40 w-100p ">Inclusive of all taxes </p>
										</div>
										<Button size="small">Book</Button>
									</div>
									<div>
										<ul>
											<li className="flex flex-middle">
												<Breakfast fill="#333333" />
												<span className="h-40 flex flex-middle pt-10 w-100p pb-10 bb fs-13 c-black-90 ml-10" style={{ borderColor: 'rgb(0,0,0,0.1)', borderWidth: '0.5px' }}>
													Breakfast
                        						</span>
											</li>
											<li className="flex flex-middle">
												<Taxes fill="#333333" />
												<span className="h-40 flex flex-middle pt-10 w-100p pb-10 bb fs-13 c-black-90 ml-10" style={{ borderColor: 'rgb(0,0,0,0.1)', borderWidth: '0.5px' }}>
													Includes all applicable taxes
                        						</span>
											</li>
											<li className="flex flex-middle" style={{ paddingLeft: '14px' }}>
												<span className="h-40 flex flex-middle pt-10 w-100p pb-10 fs-13 c-light-green ml-10">
													Refundable
													<InfoIcon
														className="ml-5"
														content={
															['Free cancellation is eligble until 12 Dec',
																'ontrary to popular belief, Lorem Ipsum is not simply random text. It has roots in a piece of classical Latin literature from 45 BC, making it over 2000 years old']
														}
													/>
												</span>
											</li>
										</ul>
									</div>
								</div>
							</div>
						</ListItem>
						<Link to="/m/v2/hotels/details/rates">
							<ListItem className="bb-0" arrow={true}>
								<a className="fs-17 bt bb c-blue h-48 flex flex-middle">See all rooms and rates</a>
							</ListItem>
						</Link>
					</List>

					{/* Similar Hotels */}
					<div className="mt-30 mb-40">
						<h5 className="fs-18 fw-600 c-black-80 mb-10 pl-16 pr-16">
							Similar hotels
						</h5>
						<Slider>
							<HotelsCard
								imgUrl={HotelFacade2}
								title="Holiday Inn Resort Goa"
								price="₹ 12,400"
								actualAmount="₹ 13,400"
							/>
							<HotelsCard
								imgUrl={HotelImage1}
								title="Holiday Inn Resort Goa"
								price="₹ 12,400"
								actualAmount="₹ 13,400"
							/>
							<HotelsCard
								imgUrl={HotelFacade2}
								title="Holiday Inn Resort Goa"
								price="₹ 12,400"
								actualAmount="₹ 13,400"
							/>
							<HotelsCard
								imgUrl={HotelFacade2}
								title="Holiday Inn Resort Goa"
								price="₹ 12,400"
								actualAmount="₹ 13,400"
							/>
						</Slider>

					</div>
					<Sticky position="bottom">
						<div className="flex flex-nowrap pt-15 pb-15 pr-16 pl-16 bg-white bt"
							style={{ borderColor: 'rgb(194,197,204)' }}>
							<Button
								className="mt-16 mb-16 pr-16 pl-16"
							>
								Select room
																							</Button>

						</div>
					</Sticky>
				</Scroller>
			</div>
		)
	}
}

export default Results
