// ct-pwa/src/components/Trains/Confirmation/index.jsx
import React, { Component } from 'react'
import PropTypes from 'prop-types'
import { browserHistory } from 'Utils/history'

import Button from 'Lib/Buttons'
import { List, ListItem } from 'Lib/List'

import { TRAINS_BASE_PATH } from 'Constants/trains/paths'

import Badge from './images/confirmation_badge@2x.png'
import IndianRupee from './icons/rupee-indian.svg'

class Confirmation extends Component {
  _redirectToHome = () => {
    browserHistory.push({
      pathname: TRAINS_BASE_PATH
    })
  }
  render() {
    return (
      <div>
        <List className="mt-10">
          <ListItem className="h-96 flex flex-middle">
            <div className="w-80p">
              <h5 className="fw-600 fs-20 c-black-90 mb-5">
                Booking confirmed
              </h5>
              <p className="fs-15 c-black-70 lh-title">
                Your trip to {this.props.destination} is booked.
              </p>
            </div>
            <img
              src={Badge}
              width="65"
              alt="Booking confirmed"
              style={{ position: 'absolute', right: '10px' }}
            />
          </ListItem>
          <ListItem className="h-48 flex flex-middle flex-between">
            <p className="fs-15 c-black-90">Paid</p>
            {this.props.fare && (
              <p className="fs-15 c-black-90">
                <IndianRupee fill="rgba(0, 0, 0, 0.9)" /> {this.props.fare}
              </p>
            )}
          </ListItem>
          <ListItem className="h-48 flex flex-middle flex-between">
            <p className="fs-15 c-black-90">Trip ID</p>
            <p className="fs-15 c-black-90">{this.props.tripId}</p>
          </ListItem>
        </List>
        <div className="mt-30 pl-16 pr-16">
          <Button type="secondary" onClick={this._redirectToHome}>
            Back to home
          </Button>
        </div>
      </div>
    )
  }
}

Confirmation.propTypes = {
  tripId: PropTypes.string,
  fare: PropTypes.string,
  destination: PropTypes.string
}

Confirmation.defaultProps = {
  tripId: '',
  destination: '',
  fare: ''
}

export default Confirmation
