// ct-pwa/src/components/Hotels/Results/HotelItem/index.jsx
import React from 'react'
import PropTypes from 'prop-types'
import range from 'lodash.range'

import { Truncate } from 'Lib/Text'
import { formatCurrency } from 'Utils'
import { formatMeasurementUnit } from 'Utils/hotels'
import { TripAdvisor } from 'Lib/Rating'

import Star from './icons/star.svg'
import FreeIcon from './icons/free.svg'

const CURRENCY = 'INR'

const HotelItem = props => {
  const imageStyles = {
    width: '104px',
    height: '114px',
    flexShrink: '0',
    backgroundSize: 'cover',
    backgroundColor: '#eaeaea',
    backgroundImage: `url("${props.image}")`,
    borderRadius: '2px'
  }

  const infoStyles = {
    width: 'calc(100% - 114px)'
  }
  const taLogoSize = {
    width: 19,
    height: 11
  }
  const reviewLogoSize = {
    width: 11,
    height: 11
  }
  const roundedRating = Math.floor(props.rating)
  const showHalfRating = props.rating !== roundedRating
  return (
    <div className="flex flex-between pt-12 pb-12">
      <div style={imageStyles} />
      <div className="ml-10 flex flex-between flex-column" style={infoStyles}>
        <div>
          <Truncate className="fs-16 fw-500 c-black lh-title mb-2">
            {props.name}
          </Truncate>
          <div className="flex flex-middle t-truncate">
            <p className="fs-11 c-black-40 mr-5 ">{props.location}</p>
            <If condition={props.distanceFromCurrentLocation}>
              <p className="fs-11 c-blue fw-600 mr-5">·</p>
              <p className="fs-11 c-blue fw-600">
                {formatMeasurementUnit(Number(props.distanceFromCurrentLocation))}
              </p>
            </If>
          </div>
        </div>
        <div className="flex flex-middle">
          <div className="w-60p">
            <div className="flex flex-column">
              <div className="d-ib flex">
                {range(props.starRating).map(i => {
                  return <Star fill="#fec22d" width="12" height="12" />
                })}

              </div>

              <div className="mt-2">
                <If condition={props.tripAdvisorRating}>
                  <div className="d-ib flex flex-middle">
                    <TripAdvisor
                      showRounderRating={false}
                      rating={props.tripAdvisorRating}
                      reviewLogoSize={reviewLogoSize}
                      taLogoSize={taLogoSize}
                    />
                  </div>
                  <span className="mr-5 fs-14">
                    {`${props.tripAdvisorRating.toFixed(1)}`}
                  </span>
                  <If condition={props.tripAdvisorReviewCount}>
                    <p className="fs-11 c-black-40 mt-3 d-ib">
                      <If condition={props.tripAdvisorReviewCount > 0}>
                        {props.tripAdvisorReviewCount} reviews
                      </If>
                    </p>
                  </If>
                </If>

              </div>
            </div>

          </div>


          <div className="flex flex-column flex-center ta-r w-40p">
            {/* <If condition={props.deal}>
              <p className="fs-12 fw-600 c-red lh-title mb-2">
                {props.discountPercentage}% OFF
              </p>
            </If> */}
            <If condition={props.deal}>
              <p className="fs-11 c-black-40 lh-title mb-2">
                <strike>{formatCurrency(Number(props.strikedOutPrice))}</strike>
              </p>
            </If>
            <p className="fs-15 fw-500 c-black-80 lh-title mb-2">
              {formatCurrency(Number(props.mrp))}
            </p>
            <If condition={props.tax}>
              <p className="fs-11 c-black-40">
                + {formatCurrency(Number(props.tax))} taxes
            </p>
            </If>

            {/* <If condition={props.payAtHotel}>
              <p className="fs-14 fw-600 c-red pt-10">
                <PAHIcon fill="#828acd" width="40" height="40 " />
              </p>
            </If> */}

            {/* <If condition={props.hotelWithDeals}>
              <p className="fs-14 fw-600 c-red pt-10">
                <DealsIcon fill="#f86810" width="24" height="24" />
              </p>
            </If> */}
          </div>
        </div>
        <If condition={props.freeCancellation}>
          <div className="flex flex-between">
            <div className="flex flex-center flex-middle">
              {/* <FreeIcon
                fill="#56953e"
                width="13"
                height="13"
                className="mr-5"
              /> */}
              <span className="fs-11 small:fs-8 c-green">
                Free cancellation
              </span>
            </div>

            <span className="fs-11 c-black">
              Total {formatCurrency(Number(props.discountedPrice))}
            </span>
          </div>
        </If>
        <If condition={!props.freeCancellation}>
          <div className="flex flex-between">
            {/* <div className="flex flex-middle"> */}
            <span className="fs-11 small:fs-8 c-black-40">
              Cancellation charges apply
              </span>
            <span className="fs-11 c-black">
              Total {formatCurrency(Number(props.discountedPrice))}
            </span>
            {/* </div> */}
          </div>
        </If>
      </div>
    </div>
  )
}

HotelItem.propTypes = {
  name: PropTypes.string.isRequired,
  location: PropTypes.bool,
  payAtHotel: PropTypes.bool,
  freeCancellation: PropTypes.bool,
  hotelWithDeals: PropTypes.bool,
  discountPercentage: PropTypes.number,
  distanceFromCurrentLocation: PropTypes.number,
  deal: PropTypes.bool,
  image: PropTypes.string,
  discountedPrice: PropTypes.number,
  tripAdvisorReviewCount: PropTypes.number,
  strikedOutPrice: PropTypes.number,
  starRating: PropTypes.number,
  tripAdvisorRating: PropTypes.number
}

HotelItem.defaultProps = {
  location: false,
  payAtHotel: false,
  freeCancellation: false,
  hotelWithDeals: false,
  discountPercentage: 0,
  distanceFromCurrentLocation: null,
  deal: false,
  image: '',
  discountedPrice: 0,
  tripAdvisorReviewCount: 0,
  strikedOutPrice: 0,
  starRating: 0,
  tripAdvisorRating: 0
}

export default HotelItem
