// ct-pwa/src/lib/FabDefault/index.jsx
import React, { PureComponent } from "react";
import PropTypes from "prop-types";
import classnames from "classnames";

import Ripple from "Lib/Animate/Ripple/";
import FabActiveIcon from "./icons/fab-active.svg";

class Fab extends PureComponent {
  render() {
    const {
      positionFlag, 
      className, 
      disabled, 
      icon, 
      label, 
      filter,
      style
    } = this.props

    const classes = positionFlag ? "Fab Fab-default" : "Fab"
    const btnClass = classnames(classes, className)
    const isDisabled = disabled
    
    return (
      <Ripple enableRipple={ this.props.enableRipple }>
        {(rippleProps, RippleContent) => {
          return (
			    <button
              disabled= { isDisabled }
              className= { btnClass }
              onClick= { this.props.onClick }
              style={{
                right:'8px',
                ...style
              }}
              { ...rippleProps }
            >
              {filter ? 
                <FabActiveIcon
                    width="20"
                    height="20"
                    fill="#339900"
                    className="Fab__icon--active"
                    style={{
                      right: '4px'
                    }}
                    />
                : null
              }
              {icon}
              {label ? 
                <span className="ml-5 fs-14">
                  {label}
                </span>
              :null
              }
              {RippleContent}
            </button>
          );
        }}
      </Ripple>
    );
  }
}

Fab.propTypes = {
  className: PropTypes.string,
  filter: PropTypes.bool,
  onClick: PropTypes.func,
  enableRipple: PropTypes.bool,
  positionFlag: PropTypes.bool,
  icon: PropTypes.element,
  label: PropTypes.string
};

Fab.defaultProps = {
  filter: false,
  className: "bg-darker-gray",
  onClick: () => {},
  enableRipple: true,
  positionFlag: false,
  icon: '',
  label: ""
};

export default Fab;
