// ct-pwa/src/lib/Pageheader/index.jsx
import React from 'react'
import PropTypes from 'prop-types'
import classNames from 'classnames'

import { Truncate } from 'Lib/Text'
import { pathType } from 'Constants/types'
import { goBack } from 'Utils/history'

import Back from './icons/back.svg'

const redirectToHome = event => {
  window.location.href = '/'
}

const Pageheader = ({
  title,
  subtitle,
  backButton,
  className,
  backNavigationParams
}) => {
  const handleBackNavigation = () => {
    goBack(backNavigationParams)
  }

  return (
    <section className={"pr-15 pl-15 flex flex-middle ta-c bb flex-center bg-white "+classNames(className)} style={{
      borderColor: '#ddd',
      height: '48px'
    }}>
      <If condition={backButton}>
        <div>
          <div onClick={handleBackNavigation} className="">
            <Back fill="#3366cc" />
          </div>
        </div>
      </If>
      <div style={{
        width: 'calc(100% - 13px)'
      }}>
        <div>
          <Truncate className="fw-600 fs-15 d-ib lh-title c-black-80">
            {title}
          </Truncate>
          <Truncate className="fs-11 lh-title c-black-40">{subtitle}</Truncate>
        </div>
      </div>
    </section>
  )
}

Pageheader.propTypes = {
  title: PropTypes.node.isRequired,
  subtitle: PropTypes.node.isRequired,
  backButton: PropTypes.bool,
  backNavigationParams: pathType
}

Pageheader.defaultProps = {
  backButton: true,
  backNavigationParams: {}
}

Pageheader.displayName = 'Pageheader'

export default Pageheader
