// ct-pwa/src/lib/Modal/index.jsx
import React, { PureComponent } from 'react'
import PropTypes from 'prop-types'
import classnames from "classnames";
import Info from './icons/info.svg'
import Warning from './icons/warning.svg'

class Modal extends PureComponent {
  _handlePrimaryAction = () => {
    this.props.onPrimaryAction()
  }

  _handleSecondaryAction = () => {
    this.props.onSecondaryAction()
  }

  render() {
    const { type, primaryButton, secondaryButton, bottom, className } = this.props
    const classes = classnames(
      className
    )
    return (
      <div className={"Modal " + (bottom ? "w-100p p-absolute b-0 Modal__bottom " : "") + classes}>
        <div className={"Modal__content " + (type === 'renderProp' ?  'pt-0 pb-0 pl-0 pr-0' : '')} >
          <Choose>
            <When condition={type === 'info'}>
              <Info width="26px" height="26px" fill="#36c" className="mb-15" />
            </When>
            <When condition={type === 'warning'}>
              <Warning
                width="26px"
                height="26px"
                fill="#36c"
                className="mb-15"
              />
            </When>
          </Choose>
          {type !== 'renderProp'?
            <h1 className="Modal__heading">{this.props.title}</h1>
            :''
          }
          <Choose>
              <When condition={type==='renderProp'}>
                {this.props.children}
              </When>
              <Otherwise>
                  <p className="Modal__copy">{this.props.message}</p>
              </Otherwise>
          </Choose>
        </div>
        {
          !bottom ?
          <div className="Modal__actions">
          <If condition={primaryButton}>
            <a
              className="Modal__action Modal__action--primary"
              onClick={this._handlePrimaryAction}
            >
              {primaryButton}
            </a>
          </If>
          <If condition={secondaryButton}>
            <a className="Modal__action" onClick={this._handleSecondaryAction}>
              {secondaryButton}
            </a>
          </If>
        </div> 
        : ''
        }
      </div>
    )
  }
}

Modal.propTypes = {
  onPrimaryAction: PropTypes.func,
  onSecondaryAction: PropTypes.func,
  title: PropTypes.string,
  message: PropTypes.string,
  type: PropTypes.oneOf(['info', 'warning', 'error','renderProp']),
  primaryButton: PropTypes.string,
  bottom: PropTypes.boolean,
  secondaryButton: PropTypes.string,
}

Modal.defaultProps = {
  type: 'info',
  onPrimaryAction: () => {},
  onSecondaryAction: () => {},
  title: '',
  message: '',
  primaryButton: '',
  secondaryButton: '',
  bottom : false
}

export default Modal
