// ct-pwa/src/constants/googleUrls/index.js
import { isEmpty, toQueryString } from 'Utils'
import format from 'date-fns/format'
import differenceInDays from 'date-fns/differenceInDays'

export const dateFormatHPA = 'YYYY-MM-DD'

export const getGoogleHPAConversionURLForHotels = (
  confirmationData,
  HOTELS
) => {
  const {
    tripId,
    totalAmount,
    baseAmount,
    hotel: { id },
    itinerary: {
      searchCriteria: { checkInDate, checkOutDate }
    }
  } = confirmationData
  const checkInDateObject = new Date(parseInt(checkInDate))
  const checkOutDateObject = new Date(parseInt(checkOutDate))
  const URL = '//www.googletraveladservices.com/travel/clk/pagead/conversion'
  const googleCustomParams = {
    hct_partner_hotel_id: id,
    hct_base_price: parseFloat(baseAmount.replace(/,/g, '')),
    hct_total_price: parseFloat(totalAmount.replace(/,/g, '')),
    hct_currency_code: 'INR',
    hct_checkin_date: format(checkInDateObject, dateFormatHPA),
    hct_checkout_date: format(checkOutDateObject, dateFormatHPA),
    hct_length_of_stay: differenceInDays(checkOutDateObject, checkInDateObject),
    hct_date_format: "%Y-%m-%d",
    hct_booking_xref: tripId,
    hct_ver: HOTELS.HPA_HCT_VER
  }
  const dataQuery = toQueryString(googleCustomParams, ';', '=')
  const fixedData = {
    label: 'HPA',
    guid: 'ON',
    script: 0,
    ord: new Date().getTime()
  }
  const fixedDataQuery = toQueryString(fixedData, '&', '=')
  const IMAGE_URL = `${URL}/${
    HOTELS.HPA_CONVERSION_ID
  }/?${fixedDataQuery}&data=${dataQuery}`
  return IMAGE_URL
}


export const getGoogleHPAConversionURLForHotelsPull = (
  confirmationData,
  HOTELS
) => {
  const {
    tripId,
    totalAmount,
    baseAmount,
    hotel: { id },
    itinerary: {
      searchCriteria: { checkInDate, checkOutDate }
    }
  } = confirmationData
  const checkInDateObject = new Date(parseInt(checkInDate))
  const checkOutDateObject = new Date(parseInt(checkOutDate))
  const URL = '//www.googletraveladservices.com/travel/clk/pagead/conversion'
  const googleCustomParams = {
    hct_partner_hotel_id: id,
    hct_base_price: parseFloat(baseAmount.replace(/,/g, '')),
    hct_total_price: parseFloat(totalAmount.replace(/,/g, '')),
    hct_currency_code: 'INR',
    hct_checkin_date: format(checkInDateObject, dateFormatHPA),
    hct_checkout_date: format(checkOutDateObject, dateFormatHPA),
    hct_length_of_stay: differenceInDays(checkOutDateObject, checkInDateObject),
    hct_date_format: "%Y-%m-%d",
    hct_booking_xref: tripId,
    hct_ver: HOTELS.HPA_HCT_VER
  }
  const dataQuery = toQueryString(googleCustomParams, ';', '=')
  const fixedData = {
    label: 'HPA',
    guid: 'ON',
    script: 0,
    ord: new Date().getTime()
  }
  const fixedDataQuery = toQueryString(fixedData, '&', '=')
  const IMAGE_URL = `${URL}/${
    HOTELS.HPA_CONVERSION_PULL_ID
  }/?${fixedDataQuery}&data=${dataQuery}`
  return IMAGE_URL
}
