// ct-pwa/src/utils/hotels/cancellationPolicy.js
import axios from "axios";
import { toQueryString, getLanguageMeta } from "Utils";
import { HOTELS_ROOM_CANCELLATION_POLICY } from "Constants/api/hotels";

export default async function getCancellationPolicy(inputData) {
  const {
    roomtypecode,
    chkOutDate,
    chkInDate,
    city,
    cnm,
    state,
    bookingCode,
    travellers,
    ct_hotelid,
    searchInstanceID
  } = inputData;

  console.log(inputData);

  const payloadData = {
    bookingCode: `${bookingCode}`,
    roomTypeCode: roomtypecode,
    state,
    city,
    cnm,
    country: getLanguageMeta().country,
    ct_hotelid,
    num_children: travellers.childs,
    num_adults: travellers.adults,
    chk_in: chkInDate,
    chk_out: chkOutDate,
    ...travellers
  };
  const payload = toQueryString(payloadData);

  console.log(payload);

  const responseData = await axios.get(`${HOTELS_ROOM_CANCELLATION_POLICY}?${payload}`, {
    headers: {
      Accept:
        "text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,image/apng,*/*;q=0.8,application/signed-exchange;v=b3",
      "Content-Type": "application/x-www-form-urlencoded"
    }
  });

  return responseData;
}
