// ct-pwa/src/lib/Buttons/ShortlistButton.jsx
import React, { PureComponent } from 'react'
import PropTypes from 'prop-types'
import classnames from 'classnames'

import Button from './Button'

import Heart from './icons/heart.svg'
import HeartOutline from './icons/heart_outline.svg'


class ShortlistButton extends PureComponent {
    _onClick = event => {
        this.props.onClick(event)
    }

    render() {
        const { isShortlisted, className } = this.props
        const btnClass = classnames(
            'pr-16',
            'pl-16',
            'Button--outline',
             className
            )

        return (
            <Button className={btnClass} onClick={this._onClick}>
                {isShortlisted 
                ?  
                <div className="flex flex-middle flex-center">
                    <Heart fill="#CC3333" className="mr-8"/> <span>Shortlisted</span>
                </div>
                : 
                <div className="flex flex-middle flex-center">
                    <HeartOutline  className="mr-8"/><span>Shortlist</span>
                </div>
                }
            </Button>
        )
    }
}

ShortlistButton.propTypes = {
    isShortlisted: PropTypes.bool,
    children: PropTypes.string,
    className: PropTypes.string,
    onClick: PropTypes.func
}

ShortlistButton.defaultProps = {
    isShortlisted: false,
    children: 'Shortlist',
    className: '',
    onClick: () => { }
}

export default ShortlistButton
