// ct-pwa/src/components/Hotels/Travellers/index.jsx
import React, { PureComponent } from 'react'
import InputBlock from 'Lib/Form/InputBlock'
import PropTypes from 'prop-types'

class Travellers extends PureComponent {
  state = {}

  travellersDetails = () => {
    const { roomsCount, adultsCount, childrenCount } = this.props
    const childCount = `, ${childrenCount} ${
      childrenCount === 1 ? 'child' : 'children'
    }`
    const adultCount = `${adultsCount} ${
      adultsCount === 1 ? 'adult' : 'adults'
    }`

    const roomCount = `${roomsCount} ${roomsCount === 1 ? 'room' : 'rooms'}`
    return `${roomCount}, ${adultCount}${childrenCount ? childCount : ''}`
  }

  render() {
    return (
      <InputBlock
        label="Travellers"
        fieldClass="TouchAndGo__tappable--row"
        labelClass="fs-15 mb-10"
        valueClass="fs-18 fw-500 c-black-80"
        value={this.travellersDetails()}
        onClick={this.props.toggleTravellerSheet}
      />
    )
  }
}

Travellers.propTypes = {
  toggleTravellerSheet: PropTypes.func,
  adultsCount: PropTypes.number,
  childrenCount: PropTypes.number,
  roomsCount: PropTypes.number
}

Travellers.defaultProps = {
  toggleTravellerSheet: () => {},
  adultsCount: 1,
  childrenCount: 0,
  roomsCount: 0
}

export default Travellers
