// ct-pwa/config/settings.js
module.exports = {
    production: {
      GOOGLE_MAPS_API_KEY: JSON.stringify(
        'AIzaSyBICejdTstEMJamaR3_1yVvohjIhLikbCM'
      ),
      GOOGLE_STATIC_MAPS_API_KEY: JSON.stringify(
        'AIzaSyBICejdTstEMJamaR3_1yVvohjIhLikbCM'
      ),
      SENTRY_DSN: JSON.stringify(
        'https://33b33adfe6264aa5ba796e334659278c@sentry.io/223228'
      ),
      CLEVERTAP_KEY: JSON.stringify('W8R-KK8-W74Z'),
      FACEBOOK_PIXEL_ID: JSON.stringify('288042424732696'),
      FLIGHTS_CONFIRMATION_FB_PIXEL_ID: JSON.stringify('6016136237563'),
      BRANCH_PIXEL_KEY: JSON.stringify(
        'key_live_cgeDODFZsneCcLhrQ9wxGomiEqkFtGo9'
      ),
      GTM_ID: JSON.stringify('GTM-T9S432'),
      DEFAULT_GA_ACCOUNT_ID: JSON.stringify('UA-8292447-1'),
      OM: {
        GTM_ID: JSON.stringify('GTM-P6JVCQ2'),
        GA_ACCOUNT_ID: JSON.stringify('UA-16109308-2'),
        CONVERSION_ID: 1000705187,
        CONVERSION_LABEL: JSON.stringify('2qnOCOW6yQMQo5mW3QM')
      },
      QA: {
        GTM_ID: JSON.stringify('GTM-P6JVCQ2'),
        GA_ACCOUNT_ID: JSON.stringify('UA-16109308-4'),
        CONVERSION_ID: 1002197040,
        CONVERSION_LABEL: JSON.stringify('9OlkCJCvlwMQsKDx3QM')
      },
      BH: {
        GTM_ID: JSON.stringify('GTM-P6JVCQ2'),
        GA_ACCOUNT_ID: JSON.stringify('UA-16109308-3'),
        CONVERSION_ID: 1005030997,
        CONVERSION_LABEL: JSON.stringify('NvdCJvbnAMQ1Zye3wM')
      },
      KW: {
        GTM_ID: JSON.stringify('GTM-P6JVCQ2'),
        GA_ACCOUNT_ID: JSON.stringify('UA-16109308-5'),
        CONVERSION_ID: 1002172652,
        CONVERSION_LABEL: JSON.stringify('HpyPCLTUwwMQ7OHv3QM')
      },
      ME: {
        GTM_ID: JSON.stringify('GTM-T9S432'),
        GA_ACCOUNT_ID: JSON.stringify('UA-8292447-22'),
        CONVERSION_ID: 1002172652,
        CONVERSION_LABEL: JSON.stringify('HpyPCLTUwwMQ7OHv3QM')
      },
      SA: {
        GTM_ID: JSON.stringify('GTM-T96QGPF'),
        GA_ACCOUNT_ID: JSON.stringify('UA-127559303-1'),
        CONVERSION_ID: 1004330979,
        CONVERSION_LABEL: JSON.stringify('qyC5CPWiygMQ47_z3gM')
      },
      AE: {
        GTM_ID: JSON.stringify('GTM-P6JVCQ2'),
        GA_ACCOUNT_ID: JSON.stringify('UA-16109308-1'),
        CONVERSION_ID: 1001005067,
        INTL_CONVRESION_LABEL: JSON.stringify('vtciCKXxwwcQi8Co3QM'),
        DOM_CONVERSION_LABEL: JSON.stringify('Cg2TCK3wwwcQi8Co3QM'),
        IMG_CONVERSION_ID: 1019108746,
        IMG_CONVERSION_LABEL: JSON.stringify('EdFdCObgqgMQirv55QM')
      },
      IN: {
        GA_ACCOUNT_ID: JSON.stringify('UA-8292447-1'),
        CONVERSION_ID: 920937462,
        CONVERSION_LABEL: JSON.stringify('AB-dCK_qlGYQ9seRtwM'),
        CONVERSION_VALUE: 1,
        OLD_IMG_CONVERSION_LABEL: JSON.stringify('C7i7CJmNxAMQs5aC_QM'),
        NEW_IMG_CONVERSION_LABEL: JSON.stringify('6iuWCJutnwYQzZn33QM'),
        INTL_DOM_CONVERSION_ID: 995874483,
        INTL_CONVERSION_LABEL: JSON.stringify('hODQCK3zuwMQs63v2gM'),
        DOM_CONVERSION_LABEL: JSON.stringify('r5DoCL3xuwMQs63v2gM'),
        NEW_IMG_CONVERSION_ID: 1002294477
      },
      FLIGHTXP: {
        GTM_ID: JSON.stringify('GTM-WCG8ZCD')
      },
      FLIGHTS: {
        MCC_CONVERSION_ID: 1067486003
      },
      HOTELS: {
        MCC_CONVERSION_ID: 992967773,
        MCC_CONVERSION_LABEL: JSON.stringify('EsoUCN3dj4QBEN34vdkD'),
        ACCOUNT_CONVERSION_ID: 994992302,
        ACCOUNT_CONVERSION_LABEL: JSON.stringify('m_ifCKrW9wEQrsG52gM'),
        HPA_CONVERSION_ID: 248,
        HPA_CONVERSION_PULL_ID: 1562831241,
        HPA_CONVERSION_LABEL: JSON.stringify('HPA'),
        HPA_CONVERSION_LANGUAGE: JSON.stringify('en'),
        HPA_CONVERSION_FORMAT: 3,
        HPA_CONVERSION_COLOR: JSON.stringify('ffffff'),
        HPA_HCT_VER: JSON.stringify('1.0.j')
      },
      IMG1_CONVERSION_ID: 990047230,
      IMG1_CONVERSION_LABEL: JSON.stringify('3-fCCOquqwcQ_teL2AM'),
      IMG2_CONVERSION_ID: 986824040,
      IMG2_CONVERSION_LABEL: JSON.stringify('MQZZCPiz3wYQ6PrG1gM'),
      IMG3_CONVERSION_ID: 985850841,
      IMG3_CONVERSION_LABEL: JSON.stringify('kNs7CKe2_AYQ2ceL1gM'),
      INTL_CONVERSION_LABEL: JSON.stringify('XVkiCJXw4wcQs63v2gM'),
      DOM_IMG1_CONVERSION_LABEL: JSON.stringify('3z9TCJ3v4wcQs63v2gM'),
      DOM_IMG2_CONVERSION_LABEL: JSON.stringify('qFL1CO304wcQs63v2gM'),
      GOOGLE_CONVERSION_COLOR: JSON.stringify('ffffff'),
      GOOGLE_CONVERSION_LANGUAGE: JSON.stringify('en')
    },
    qa: {
      GOOGLE_MAPS_API_KEY: JSON.stringify(
        'AIzaSyBICejdTstEMJamaR3_1yVvohjIhLikbCM'
      ),
      GOOGLE_STATIC_MAPS_API_KEY: JSON.stringify(
        'AIzaSyBICejdTstEMJamaR3_1yVvohjIhLikbCM'
      ),
      SENTRY_DSN: JSON.stringify(
        'https://8b4cc49d1a0a467daa08f64a2a63a96b@sentry.io/221522'
      ),
      CLEVERTAP_KEY: JSON.stringify('TEST-8WW-9K9-W94Z'),
      FACEBOOK_PIXEL_ID: JSON.stringify('288042424732696'),
      FLIGHTS_CONFIRMATION_FB_PIXEL_ID: JSON.stringify('6016136237563'),
      BRANCH_PIXEL_KEY: JSON.stringify(
        'key_live_cgeDODFZsneCcLhrQ9wxGomiEqkFtGo9'
      ),
      GTM_ID: JSON.stringify('GTM-T9S432'),
      DEFAULT_GA_ACCOUNT_ID: JSON.stringify('UA-8292447-1'),
      OM: {
        GTM_ID: JSON.stringify('GTM-P6JVCQ2'),
        GA_ACCOUNT_ID: JSON.stringify('UA-16109308-2'),
        CONVERSION_ID: 1000705187,
        CONVERSION_LABEL: JSON.stringify('2qnOCOW6yQMQo5mW3QM')
      },
      QA: {
        GTM_ID: JSON.stringify('GTM-P6JVCQ2'),
        GA_ACCOUNT_ID: JSON.stringify('UA-16109308-4'),
        CONVERSION_ID: 1002197040,
        CONVERSION_LABEL: JSON.stringify('9OlkCJCvlwMQsKDx3QM')
      },
      BH: {
        GTM_ID: JSON.stringify('GTM-P6JVCQ2'),
        GA_ACCOUNT_ID: JSON.stringify('UA-16109308-3'),
        CONVERSION_ID: 1005030997,
        CONVERSION_LABEL: JSON.stringify('NvdCJvbnAMQ1Zye3wM')
      },
      KW: {
        GTM_ID: JSON.stringify('GTM-P6JVCQ2'),
        GA_ACCOUNT_ID: JSON.stringify('UA-16109308-5'),
        CONVERSION_ID: 1002172652,
        CONVERSION_LABEL: JSON.stringify('HpyPCLTUwwMQ7OHv3QM')
      },
      ME: {
        GTM_ID: JSON.stringify('GTM-T9S432'),
        GA_ACCOUNT_ID: JSON.stringify('UA-8292447-22'),
        CONVERSION_ID: 1002172652,
        CONVERSION_LABEL: JSON.stringify('HpyPCLTUwwMQ7OHv3QM')
      },
      SA: {
        GTM_ID: JSON.stringify('GTM-T96QGPF'),
        GA_ACCOUNT_ID: JSON.stringify('UA-127559303-1'),
        CONVERSION_ID: 1004330979,
        CONVERSION_LABEL: JSON.stringify('qyC5CPWiygMQ47_z3gM')
      },
      AE: {
        GTM_ID: JSON.stringify('GTM-P6JVCQ2'),
        GA_ACCOUNT_ID: JSON.stringify('UA-16109308-1'),
        CONVERSION_ID: 1001005067,
        INTL_CONVRESION_LABEL: JSON.stringify('vtciCKXxwwcQi8Co3QM'),
        DOM_CONVERSION_LABEL: JSON.stringify('Cg2TCK3wwwcQi8Co3QM'),
        IMG_CONVERSION_ID: 1019108746,
        IMG_CONVERSION_LABEL: JSON.stringify('EdFdCObgqgMQirv55QM')
      },
      IN: {
        GA_ACCOUNT_ID: JSON.stringify('UA-8292447-1'),
        CONVERSION_ID: 920937462,
        CONVERSION_LABEL: JSON.stringify('AB-dCK_qlGYQ9seRtwM'),
        CONVERSION_VALUE: 1,
        OLD_IMG_CONVERSION_LABEL: JSON.stringify('C7i7CJmNxAMQs5aC_QM'),
        NEW_IMG_CONVERSION_LABEL: JSON.stringify('6iuWCJutnwYQzZn33QM'),
        INTL_DOM_CONVERSION_ID: 995874483,
        INTL_CONVERSION_LABEL: JSON.stringify('hODQCK3zuwMQs63v2gM'),
        DOM_CONVERSION_LABEL: JSON.stringify('r5DoCL3xuwMQs63v2gM'),
        NEW_IMG_CONVERSION_ID: 1002294477
      },
      FLIGHTXP: {
        GTM_ID: JSON.stringify('GTM-WCG8ZCD')
      },
      FLIGHTS: {
        MCC_CONVERSION_ID: 1067486003
      },
      HOTELS: {
        MCC_CONVERSION_ID: 992967773,
        MCC_CONVERSION_LABEL: JSON.stringify('EsoUCN3dj4QBEN34vdkD'),
        ACCOUNT_CONVERSION_ID: 994992302,
        ACCOUNT_CONVERSION_LABEL: JSON.stringify('m_ifCKrW9wEQrsG52gM'),
        HPA_CONVERSION_ID: 248,
        HPA_CONVERSION_PULL_ID: 1562831241,
        HPA_CONVERSION_LABEL: JSON.stringify('HPA'),
        HPA_CONVERSION_LANGUAGE: JSON.stringify('en'),
        HPA_CONVERSION_FORMAT: 3,
        HPA_CONVERSION_COLOR: JSON.stringify('ffffff'),
        HPA_HCT_VER: JSON.stringify('1.0.j')
      },
      IMG1_CONVERSION_ID: 990047230,
      IMG1_CONVERSION_LABEL: JSON.stringify('3-fCCOquqwcQ_teL2AM'),
      IMG2_CONVERSION_ID: 986824040,
      IMG2_CONVERSION_LABEL: JSON.stringify('MQZZCPiz3wYQ6PrG1gM'),
      IMG3_CONVERSION_ID: 985850841,
      IMG3_CONVERSION_LABEL: JSON.stringify('kNs7CKe2_AYQ2ceL1gM'),
      INTL_CONVERSION_LABEL: JSON.stringify('XVkiCJXw4wcQs63v2gM'),
      DOM_IMG1_CONVERSION_LABEL: JSON.stringify('3z9TCJ3v4wcQs63v2gM'),
      DOM_IMG2_CONVERSION_LABEL: JSON.stringify('qFL1CO304wcQs63v2gM'),
      GOOGLE_CONVERSION_COLOR: JSON.stringify('ffffff'),
      GOOGLE_CONVERSION_LANGUAGE: JSON.stringify('en')
    }
  }
