// ct-pwa/src/components/Payment/CVVInput.jsx
import React, { Component } from 'react'
import PropTypes from 'prop-types'
import classNames from 'classnames'

class CVVInput extends Component {
  shouldComponentUpdate(nextProps, nextState) {
    return true
  }

  render() {
    const classList = classNames(
      this.props.className,
      'input input--text input--masked fs-15'
    )
    return (
      <input
        value={this.props.value}
        type="tel"
        autoComplete={this.props.autoComplete}
        name={this.props.name}
        className={classList}
        maxLength={this.props.maxLength}
        placeholder={this.props.placeholder}
        onChange={this.props.onChange}
        onFocus={this.props.onFocus}
      />
    )
  }
}

CVVInput.propTypes = {
  value: PropTypes.string.isRequired,
  maxLength: PropTypes.number.isRequired,
  onChange: PropTypes.func.isRequired,
  className: PropTypes.string,
  placeholder: PropTypes.string,
  name: PropTypes.string.isRequired,
  onFocus: PropTypes.func,
  autoComplete: PropTypes.string
}

CVVInput.defaultProps = {
  className: '',
  placeholder: 'CVV',
  onFocus: () => {},
  autoComplete: 'cc-csc'
}

export default CVVInput
