// ct-pwa/src/components/Payment/SaveCard.jsx
import React, { PureComponent } from 'react'
import PropTypes from 'prop-types'

import { RadioList, RadioItem } from 'Lib/Radio'

class SaveCard extends PureComponent {
  render() {
    return (
      <div>
        <RadioList
            value={this.props.enableSavedCard}
            onChange={this.props.toggleSaveCard}
            className="mt-30"
            multi={true}
            >
            <RadioItem
                value="true"
                className="RadioList__item h-48 flex flex-middle"
            >
                {selected => (
                <div>
                    <p className="fs-15 c-black-80 pt-4 mr-10 ml-10 lh-title">
                    {this.props.saveCardText}
                    </p>
                    <p className="fs-12 c-black-80 pt-1 pb-4 mr-10 ml-10 lh-title">
                    {this.props.neverSaveCVVText}
                    </p>
                </div>
                )}
            </RadioItem>
        </RadioList>
      </div>
    )
  }
}

SaveCard.propTypes = {
    enableSavedCard: PropTypes.array,
    toggleSaveCard: PropTypes.func,
    saveCardText: PropTypes.string.isRequired,
    neverSaveCVVText: PropTypes.string.isRequired
}
SaveCard.defaultProps = {
    enableSavedCard: [],
    toggleSaveCard: () => {}
}

export default SaveCard
