// ct-pwa/src/components/Payment/UPI.jsx
import React, { Component } from 'react'
import classNames from 'classnames'
import PropTypes from 'prop-types'
import { List, ListItem } from 'Lib/List'
import { SectionHeader } from 'Lib/Text'
import Input from 'Lib/Form/Input'

import { RadioList, RadioItem } from 'Lib/Radio'
import Info from './icons/info@2x.png'

class UPI extends Component {
  static propTypes = {
    thirdPartyWallets: PropTypes.arrayOf(PropTypes.object),
    thirdPartyUPI: PropTypes.arrayOf(PropTypes.object),
    onWalletChange: PropTypes.func,
    selectedWallet: PropTypes.string,
    textEnterUPIID: PropTypes.string
  }

  static defaultProps = {
    thirdPartyWallets: [],
    thirdPartyUPI: [],
    onWalletChange: () => {},
    selectedWallet: '',
    textEnterUPIID: 'Enter Your UPI ID'
  }

  constructor(props) {
    super(props)
    this.state = {
      vpa: props.vpa
    }
  }
  shouldComponentUpdate(nextProps, nextState) {
    return (
      nextState.vpa !== this.state.vpa ||
      nextProps.selectedWallet !== this.props.selectedWallet
    )
  }

  _isUPIWallet = value => {
    const { thirdPartyUPI = [] } = this.props
    return thirdPartyUPI.find(upi => upi.value === value)
  }

  _handleInputChange = value => {
    this.setState({ vpa: value })
    this._handleWalletChange({ value })
  }

  _handleInputFocus = () => {
    this._handleWalletChange()
  }

  _handleWalletChange = ({ value } = {}) => {
    this.props.onWalletChange({
      walletCompany: 'RAZORPAY',
      walletType: 'up',
      vpa: value || this.state.vpa
    })
  }

  render() {
    const { textEnterUPIID, selectedWallet } = this.props
    const inputClass = classNames('fs-15 c-black-90', {
      'c-green fw-700': false,
      'ta-r': false
    })

    const couponClassName = classNames('h-48 flex flex-middle ml-25', {
      'has-error': false
    })

    return (
      <div>
        <SectionHeader>UPI</SectionHeader>
        <List className="mt-10">
          <ListItem>
            <RadioList
              className="bt-0 bb-0"
              onChange={this._handleWalletChange}
              value={selectedWallet}
            >
              <RadioItem
                className="bb-0 p-absolute ml--15 mt-15"
                value="RAZORPAY"
                key="upi_field"
              >
                {() => ''}
              </RadioItem>
            </RadioList>
            <Input
              autoComplete="off"
              type="text"
              inputClassName={inputClass}
              wrapperClassName={couponClassName}
              placeholder={textEnterUPIID}
              value={this.state.vpa}
              name="coupon"
              onChange={this._handleInputChange}
              hasError={false}
              onFocus={this._handleInputFocus}
            />
          </ListItem>
          <div
            className="flex pt-10 pb-10"
            style={{ backgroundColor: '#F9F8FB' }}
          >
            <div>
              <img src={Info} alt="Info" width={20} className="ml-15" />
            </div>
            <div className="ml-10">
              <p className="fs-12">The UPI ID format is yourname@bank</p>
              <p className="fs-12 pt-8">
                Please authorize the payment request on your UPI app
              </p>
            </div>
          </div>
        </List>
      </div>
    )
  }
}

export default UPI
