// ct-pwa/src/components/Trains/IRCTC/PrePayment/index.jsx
import React, { Component } from 'react'
import PropTypes from 'prop-types'
import { connect } from 'react-redux'

class PrePayment extends Component {
  _generateForm = () => {
    const { redirectionParams } = this.props.payment
    const {
      payment_gateway_url: PGU, // eslint-disable-line no-unused-vars
      ...paymentFormProps
    } = redirectionParams
    const serializedData = Object.keys(paymentFormProps)
    return serializedData.map(key => {
      return <input type="hidden" name={key} value={paymentFormProps[key]} />
    })
  }

  render() {
    const { formMethod, redirectionParams } = this.props.payment
    return (
      <div>
        <form
          ref={this.props.refHandler}
          method={formMethod}
          action={redirectionParams.payment_gateway_url}
        >
          {this._generateForm()}
        </form>
      </div>
    )
  }
}

PrePayment.propTypes = {
  payment: PropTypes.objectOf(
    PropTypes.oneOfType([PropTypes.string, PropTypes.object])
  ),
  refHandler: PropTypes.func
}

PrePayment.defaultProps = {
  payment: {
    status: 'PAYMENT_INITIALIZED',
    redirectionParams: {
      Merchant_Id: 'M_kar18423_18423',
      Amount: '592.00',
      Order_Id: '70059030',
      Redirect_Url: 'https://www.cleartrip.com/payment/return/ccavenue',
      Checksum: '1538531805',
      cardOption: 'netBanking',
      netBankingCards: 'UTI_N',
      billing_cust_name: 'Cleartrip',
      billing_cust_address: 'nullnullnull',
      billing_cust_state: null,
      billing_cust_country: null,
      billing_cust_city: null,
      billing_zip_code: null,
      billing_cust_tel: '9428495750',
      billing_cust_email: 'premanand@cleartrip.com',
      payment_gateway_url:
        'https://www.ccavenue.com/servlet/new_txn.PaymentIntegration',
      irctc_user_name: 'p513208099'
    },
    formMethod: 'POST',
    formAction:
      'https://www.cleartrip.com/v1/api/trains/itinerary/reset_password',
    tripId: '170822110838',
    paymentCallbackUrl: '',
    paymentTxnRef: ''
  },
  refHandler: () => {}
}

const mapStateToProps = ({ data }) => ({
  payment: data.payment.data
})

export default connect(
  mapStateToProps,
  {}
)(PrePayment)
