// ct-pwa/src/lib/AddReadMore/index.jsx
import React, { Component } from "react";
import PropTypes from "prop-types";
import classNames from "classnames";

class AddReadMore extends Component {
  constructor(props) {
    super(props);
    this.state = {
      showReadMore: false,
      truncatedString: ""
    };
  }

  componentDidMount() {
    if (this.props.children.length > this.props.charLength) {
      this.setState({ showReadMore: true });
      this.setState({
        truncatedString: this.props.children.substr(0, this.props.charLength)
      });
    } else {
      this.setState({ showReadMore: false });
      this.setState({
        truncatedString: this.props.children
      });
    }
  }

  handleClick = () => {
    this.props.onClick();
  }

  render() {
    const { className, text } = this.props;
    return (
      <p className={ `break-w ${classNames(className)}` } onClick={ this.handleClick }>
        <span>{this.state.truncatedString}</span>
        {this.state.showReadMore ? (
          <span>
            ...
            <span className="c-blue">{text}</span>
          </span>
        ) : (
          ""
        )}
      </p>
    );
  }
}
AddReadMore.propTypes = {
  className: PropTypes.string,
  text: PropTypes.string,
  charLength: PropTypes.number,
  children: PropTypes.string.isRequired
};
AddReadMore.defaultProps = {
  className: "",
  text: "",
  charLength: null,
  children: ""
};

export default AddReadMore;
