// ct-pwa/src/components/Trains/Availability/GeneralAvailablity.jsx
import React, { PureComponent } from 'react'
import PropTypes from 'prop-types'
import { ListItem } from 'Lib/List'
import { AvailabiltyBadge } from 'Lib/Badges'

class GeneralAvailability extends PureComponent {
  _handleClick = () => {
    if (this.props.status !== 'N/A' && this.props.status !== 'Closed') {
      this.props.handleCreateItinerary(
        this.props.date,
        this.props.status,
        this.props.count,
        'GN',
        this.props.activeClass
      )
    }
  }
  render() {
    return (
      <ListItem
        onClick={this._handleClick}
        className="h-48 flex flex-middle flex-between"
      >
        <p className="fs-15 c-black-90 flex-1">{this.props.formattedDate}</p>
        <p className="fs-13 c-black-40 flex-1">{this.props.class}</p>
        <AvailabiltyBadge count={this.props.count} type={this.props.status} />
      </ListItem>
    )
  }
}

GeneralAvailability.propTypes = {
  date: PropTypes.object, //eslint-disable-line
  class: PropTypes.string,
  count: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  status: PropTypes.string,
  handleCreateItinerary: PropTypes.func,
  activeClass: PropTypes.string.isRequired,
  formattedDate: PropTypes.string
}

GeneralAvailability.defaultProps = {
  class: 'General',
  handleCreateItinerary: () => {},
  count: 0,
  status: '',
  formattedDate: ''
}

export default GeneralAvailability
