// ct-pwa/src/components/Trains/BookingFailure/index.jsx
import React, { Component } from 'react'
import PropTypes from 'prop-types'
import { browserHistory } from 'Utils/history'

import Analytics from 'Utils/Analytics'
import { List, ListItem } from 'Lib/List'
import Button from 'Lib/Buttons'
import { Pagetitle } from 'Lib/Subheader'

import { TRAINS_SEARCH } from 'Constants/trains/paths'

import Warn from './warning.svg'

class BookingFailure extends Component {
  componentDidMount() {
    Analytics.event('Trains_Booking_Failure_View', {
      itinerary_id: this.props.params.id
    })
  }

  _handleTryAgain = () => {
    browserHistory.replace({
      pathName: TRAINS_SEARCH
    })
  }

  render() {
    return (
      <div>
        <Pagetitle title="Booking Falied" />
        <List className="mt-10">
          <ListItem className="pt-20 pb-20 flex flex-middle flex-center">
            <div className="ta-c">
              <Warn fill="#c30" width="30" height="30" />
              <h5 className="fw-600 fs-18 lh-title c-black-90 mb-5 mt-15">
                Oops! Your booking falied
              </h5>
              {/* eslint-disable */}
              <p className="fs-15 lh-copy c-black-70 mb-10 mt-10">
                We are trying to confirm your booking. If the confirmation does
                not happen,{' '}
                <strong className="fw-800">we will refund the amount</strong>{' '}
                you have paid.
              </p>

              <p className="fs-15 lh-copy c-black-70 mb-10 mt-10">
                Please call us in the next 15 minutes to know the latest on your
                booking.
              </p>
            </div>
          </ListItem>
        </List>
        <div className="pl-16 pr-16 mt-20 mb-20 ta-c">
          <Button size="medium" type="tertiary" onClick={this._handleTryAgain}>
            Back to search
          </Button>
        </div>
      </div>
    )
  }
}

BookingFailure.propTypes = {
  params: PropTypes.objectOf(PropTypes.any).isRequired
}

export default BookingFailure
