// ct-pwa/src/lib/Form/EmailAutoComplete.jsx
import React, { Component } from 'react'
import PropTypes from 'prop-types'
import classNames from 'classnames'

import { isEmpty } from 'Utils'
import Label from './Label'

class EmailAutoComplete extends Component {
  constructor(props) {
    super(props)
    const uniqueId = `${this.props.name}-${Math.floor(Math.random() * 0xffff)}`
    this.uniqueId = uniqueId.replace(/[^A-Za-z0-9-]/gi, '')
    this.numberRegex = /^\d+$/
    this.state = {
      value: props.value,
      emailVal: props.value || ''
    }
  }

  componentDidMount() {
    if (this.props.autoFocus) {
      setTimeout(this.focus, this.props.focusDelay)
    }
  }
  shouldComponentUpdate(nextProps, nextState) {
    return (
      this.state.value !== nextProps.value ||
      this.state.value !== this.props.value ||
      this.props.hasError !== nextProps.hasError
    )
  }

  static getDerivedStateFromProps(nextProps, prevState) {
    if (nextProps.value !== prevState.value) {
      return {
        value: nextProps.value
      }
    }
  }

  _handleChange = event => {
    const { onChange, name } = this.props
    let valueFromEvent = event.target.value
    const { domains } = this.props
    const elem = event.target
    const emailAddress = elem.value
    const { emailVal } = this.state
    if (
      domains !== undefined &&
      emailAddress.length > emailVal.length // Get suggestion only when new char is typed. Don't call for delete/backspace
    ) {
      const suggest = this._suggest(elem.value)
      const suggestedEmail = emailAddress + suggest
      if (suggest) {
        elem.setAttribute('value', suggestedEmail)
        elem.value = suggestedEmail
        const start = elem.value.lastIndexOf(suggest)
        const end = elem.value.length
        elem.setSelectionRange(start, end)
        valueFromEvent = elem.value
      }
    }
    this.setState({
      emailVal: emailAddress
    })

    // propagate to store and therefore to the input
    if (onChange) {
      this.setState({ value: valueFromEvent }, () => {
        onChange(valueFromEvent, name, event)
      })
    }
  }

  _handleKeyPress = event => {
    const { onKeyPress, type } = this.props

    if (onKeyPress) {
      onKeyPress(event)
    }
  }

  _handleOnFocus = () => {
    const { onFocus, name } = this.props

    if (onFocus) {
      onFocus(this.inputNode, name)
    }
  }

  _handleOnBlur = event => {
    this.setState({
      emailVal: event.target.value
    })
    if (window && window.getSelection) {
      window.getSelection().removeAllRanges()
    } else if (document && document.selection) {
      document.selection.empty()
    }
  }

  blur = () => {
    this.inputNode.blur()
  }

  focus = () => {
    // Temporary fix: to be modified later
    if (this.inputNode) {
      this.inputNode.focus()
    }
  }

  _suggest(string) {
    let str_arr = string.split('@')
    if (str_arr.length > 1) {
      string = str_arr.pop()
      if (!string.length) {
        return
      }
    } else {
      return
    }

    let match =
      this.props.domains
        .filter(domain => {
          return 0 === domain.indexOf(string.toLowerCase())
        })
        .shift() || ''

    return match.replace(string.toLowerCase(), '')
  }

  _renderAction = () => {
    if (this.props.actionLabel === null) return null
    return (
      <p
        onClick={this.props.onActionClick}
        className={classNames(this.props.actionClassName)}
      >
        {this.props.actionLabel}
      </p>
    )
  }

  _createInput = () => {
    let { type } = this.props
    const {
      wrapper,
      wrapperClassName,
      labelClassName, // eslint-disable-line no-unused-vars
      inputClassName,
      disabled,
      name,
      // type,
      label, // eslint-disable-line no-unused-vars
      autoFocus,
      focusDelay,
      hasError, // eslint-disable-line no-unused-vars
      id,
      autoComplete,
      placeholder,
      actionLabel,
      actionClassName,
      onActionClick,
      ...others
    } = this.props

    const { value } = this.state
    const valuePresent = !isEmpty(value)

    const inputId = id || this.uniqueId
    const labelElementProps = {
      className: classNames(labelClassName, {
        'c-red': hasError,
        'c-black-90': !hasError
      }),
      key: 'label',
      label,
      for: inputId
    }

    const inputMode = type === 'email' ? 'email' : ''
    const autocapitalize = type === 'email' ? 'off' : ''
    type = type === 'email' ? 'text' : type
    const inputElementProps = {
      ...others,
      className: classNames('input--text', inputClassName, {
        'has-value': valuePresent
      }),
      onChange: this._handleChange,
      ref: node => {
        this.inputNode = node
      },
      name,
      disabled,
      type,
      inputMode,
      autocapitalize,
      value,
      id: inputId,
      key: inputId,
      onKeyPress: this._handleKeyPress,
      onFocus: this._handleOnFocus,
      onBlur: this._handleOnBlur,
      autoComplete,
      placeholder
    }

    return [
      <Label {...labelElementProps} />,
      <input {...inputElementProps} />,
      this._renderAction()
    ]
  }

  render() {
    const { wrapper: Wrapper, wrapperClassName } = this.props
    const inputElem = this._createInput()
    if (React.isValidElement(Wrapper)) {
      return React.cloneElement(
        Wrapper,
        {
          className: classNames(wrapperClassName, {
            'has-error': this.props.hasError
          })
        },
        inputElem
      )
    }
    return (
      <Wrapper
        className={classNames(wrapperClassName, {
          'has-error': this.props.hasError
        })}
      >
        {inputElem}
      </Wrapper>
    )
  }
}

EmailAutoComplete.propTypes = {
  /**
   * Whether the input value is invalid.
   */
  hasError: PropTypes.bool,
  /**
   * Whether to enable autofocus.
   */
  autoFocus: PropTypes.bool,
  /**
   * Delay(in ms) to trigger focus on the input.
   */
  focusDelay: PropTypes.number,
  /**
   * Class attribute of the input.
   */
  inputClassName: PropTypes.string,
  /**
   * Whether the input is disabled.
   */
  disabled: PropTypes.bool,
  /**
   * ID attribute of the input.
   */
  id: PropTypes.string,
  /**
   * Label value of the input.
   */
  label: PropTypes.string,
  /**
   * Class attribute of the label.
   */
  labelClassName: PropTypes.string,
  /**
   * Name attribute of the input.
   */
  name: PropTypes.string.isRequired,
  /**
   * Event handler for on blur event of the input.
   */
  onBlur: PropTypes.func,
  /**
   * Event handler for on change event of the input.
   */
  onChange: PropTypes.func,
  /**
   * Event handler for on focus event of the input.
   */
  onFocus: PropTypes.func,
  /**
   * Event handler for on key press event of the input.
   */
  onKeyPress: PropTypes.func,
  /**
   * Whether the input is mandatory.
   */
  required: PropTypes.bool,
  /**
   * Specifies the type of the input.
   */
  type: PropTypes.string,
  /**
   * Specifies the label of action.
   */
  actionLabel: PropTypes.string,
  /**
   * Class attribute of the label.
   */
  actionClassName: PropTypes.string,
  /**
   * Specifies the label of action.
   */
  onActionClick: PropTypes.func,
  /**
   * Whether to enable browser autocomplete for the input.
   */
  autoComplete: PropTypes.string,
  value: PropTypes.oneOfType([
    // eslint-disable-line react/require-default-props
    PropTypes.number,
    PropTypes.object,
    PropTypes.string
  ]),
  wrapper: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.element,
    PropTypes.func
  ]),
  wrapperClassName: PropTypes.string,
  placeholder: PropTypes.string
}

EmailAutoComplete.defaultProps = {
  hasError: false,
  autoFocus: false,
  focusDelay: 0,
  inputClassName: '',
  disabled: false,
  id: null,
  label: '',
  labelClassName: '',
  onBlur: null,
  onChange: null,
  onFocus: null,
  onKeyPress: null,
  required: false,
  type: 'text',
  autoComplete: 'on',
  wrapper: <div />,
  wrapperClassName: '',
  actionLabel: null,
  actionClassName: '',
  onActionClick: null,
  domains: [
    'gmail.com',
    'hotmail.com',
    'icloud.com',
    'live.com',
    'msn.com',
    'outlook.com',
    'rediffmail.com',
    'yahoo.com'
  ]
}

export default EmailAutoComplete
