// ct-pwa/src/components/Hotels/Search/HotelListItem/index.jsx
import React from 'react'
import PropTypes from 'prop-types'
import range from 'lodash.range'
import classnames from 'classnames'

import Ripple from 'Lib/Animate/Ripple/'
import { TripAdvisor } from 'Lib/Rating'

import Star from './icons/star.svg'

const HotelListItem = props => {
	const imageStyles = {
		width: '80px',
		height: '80px',
		flexShrink: '0',
		backgroundSize: 'cover',
		backgroundColor: '#eaeaea',
		backgroundImage: `url("${props.image}")`,
	}
	const taLogoSize = {
		width: 21,
		height: 12
	}
	const reviewLogoSize = {
		width: 12,
		height: 12
	}
	const roundedRating = Math.floor(props.rating)
	const showHalfRating = props.rating !== roundedRating
	const { name,
		location,
		date,
		starRating,
		tripAdvisorRating,
		travelInfo,
		className } = props

	return (
		<div className={classnames('flex flex-start pt-15 pb-15', className)}>
			<div className="br2" style={imageStyles}></div>

			<div className="flex flex-column flex-between pl-12">
				<div className="flex flex-top flex-column">
					<h4 className="fs-16 fw-500 lh-title p-truncate-2 pb-4">
						{name}
					</h4>
					<If condition={location}>
						<p className="fs-12 lh-copy c-black-60">{location}</p>
					</If>
				</div>
				<div>
					<If condition={date}>
						<p className="fs-12  lh-copy">{date}</p>
					</If>
					<If condition={travelInfo}>
						<p className="fs-12  lh-copy c-black-60">{travelInfo}</p>
					</If>
				</div>

				<If condition={tripAdvisorRating || starRating}>
					<div className="flex flex-row mt-8 flex-middle">
						<If condition={tripAdvisorRating}>
							<TripAdvisor
								rating={4}
								reviewLogoSize={reviewLogoSize}
								taLogoSize={taLogoSize}
							/>
						</If>
						<If condition={starRating}>
							<div className="d-ib flex pl-16">
								{range(4).map(i => {
									return <Star fill="#fec22d" width="12" height="12" />
								})}
							</div>
						</If>
					</div>
				</If>
			</div>

		</div>
	)

}

HotelListItem.propTypes = {
	name: PropTypes.string.isRequired,
	location: PropTypes.string,
	image: PropTypes.string,
	date: PropTypes.string,
	starRating: PropTypes.number,
	tripAdvisorRating: PropTypes.number,
}

HotelListItem.defaultProps = {
	location: '',
	date: '',
	image: '',
	tripAdvisorReviewCount: 0,
	starRating: 0,
	tripAdvisorRating: 0,
}

export default HotelListItem
