// ct-pwa/src/components/Calendar/index.jsx
import React, { Component } from 'react'
import PropTypes from 'prop-types'
import TinyCalendar from 'react-tiny-calendar'

import Sheet from 'Lib/Sheet'
import { Tabs, Tab } from 'Lib/Tabs'
import { dateType } from 'Constants/types'
import startOfMonth from 'date-fns/startOfMonth'

import { startOfToday } from 'Utils/dates'

class Calendar extends Component {
  constructor(props) {
    super(props)
    this.today = startOfToday()
    this.thisMonth = startOfMonth(this.today)
  }

  handleDateTypeChange = dateType => {
    this.props.onSelectionTypeChange(dateType)
  }

  handleDateRangeSelection = selectedDates => {
    this.props.onChange(selectedDates)
  }

  render() {
    const {
      showCalendar,
      selectRange,
      minDate: minDateFromProps,
      value,
      numberOfMonths,
      onChange,
      initialMonth,
      selectionType,
      onClose
    } = this.props
    const activeStartDate = initialMonth || this.thisMonth
    const minDate = minDateFromProps || this.today
    return (
      <Sheet isOpen={showCalendar} title="Select Date" onClose={onClose}>
        <If condition={selectRange}>
          <Tabs value={selectionType} onChange={this.handleDateTypeChange}>
            <Tab value="start" label="Check In" />
            <Tab value="end" label="Check Out" />
          </Tabs>
        </If>
        <TinyCalendar
          selectRange={selectRange}
          minDate={minDate}
          value={value}
          numberOfMonths={numberOfMonths}
          onChange={this.handleDateRangeSelection}
          activeStartDate={activeStartDate}
          selectionState={selectionType}
        />
      </Sheet>
    )
  }
}

Calendar.propTypes = {
  showCalendar: PropTypes.bool,
  selectRange: PropTypes.bool,
  minDate: dateType,
  value: PropTypes.oneOfType([dateType, PropTypes.arrayOf(dateType)]),
  numberOfMonths: PropTypes.number,
  onChange: PropTypes.func,
  initialMonth: dateType,
  selectionType: PropTypes.oneOf(['start', 'end']),
  onSelectionTypeChange: PropTypes.func,
  onClose: PropTypes.func
}

Calendar.defaultProps = {
  numberOfMonths: 13,
  selectionType: 'onward'
}

export default Calendar
