// ct-pwa/src/constants/hotels/paths.js
import { appRoutePrefix } from '../'

export const HOTELS_BASE_PATH = `${appRoutePrefix}/hotels`

export const HOTELS_SEARCH_RESULTS = `${HOTELS_BASE_PATH}/results`

export const HOTEL_DETAILS = hotelId => `${HOTELS_BASE_PATH}/details/${hotelId}`

export const HOTEL_DETAILS_RATES = hotelId => `${HOTEL_DETAILS(hotelId)}/rates`

export const HOTEL_DETAILS_REVIEWS = hotelId =>
  `${HOTEL_DETAILS(hotelId)}/reviews`

export const HOTEL_DETAILS_MAP = hotelId => `${HOTEL_DETAILS(hotelId)}/map`

export const HOTELS_ITINERARY_PATH = itineraryId =>
  `${HOTELS_BASE_PATH}/itinerary/${itineraryId}/info`

export const HOTELS_TRAVELLERS_PATH = itineraryId =>
  `${HOTELS_BASE_PATH}/itinerary/${itineraryId}/traveller`

export const HOTEL_PAYMENTS_PATH = itineraryId =>
  `${HOTELS_BASE_PATH}/itinerary/${itineraryId}/payment`

export const HOTELS_PAYMENT_INTIALIZED_PATH = itineraryId =>
  `${HOTELS_BASE_PATH}/itinerary/${itineraryId}/initialize-payment`

export const HOTELS_ITINERARY_PAYMENT_RETRY_PATH = itineraryId =>
  `${HOTELS_BASE_PATH}/itinerary/${itineraryId}/payment-retry`
