// ct-pwa/src/constants/api/hotels.js
import { DOMAIN } from "../";

export const SEARCH_URL = `${DOMAIN}/m/hotels/results`;
export const AUTOCOMPLETE_HOTELS_LOCATION_URL = `${DOMAIN}/places/hac?&v=2`;
export const HOTELS_CONTENT_URL = `${DOMAIN}/hotels/jsonservice/hotel-content`;
export const HOTELS_SCDATA_URL = `${DOMAIN}/hotels/service/scData`;
export const HOTELS_SID_URL = `${DOMAIN}/hotels/service/sid-results`;

// change this to country keycode when pwa is live for other domains
export const HOTELS_SRP_MERCHANDISING = `${DOMAIN}/offermgmt/iOSB2CHotelSearch_dom_IN.json`;

// Book flow endpoints
export const HOTEL_ITINERARY = `${DOMAIN}/hotels/initiate-booking`;
export const CREATE_HOTEL_ITINERARY_URL = `${DOMAIN}/hotels/itinerary/mobile/create`;
export const HOTEL_BOOKING_POLICY_URL = itineraryId =>
  `${DOMAIN}/hotels/itinerary/${itineraryId}/bookingPolicy`;
export const HOTELS_PAYMENT_URL = itineraryId =>
  `${DOMAIN}/book/hotels/itinerary/${itineraryId}/mobile-payment`;
export const HOTELS_APPLY_COUPON_URL = itineraryId =>
  `${DOMAIN}/hotels/itinerary/${itineraryId}/apply-coupon`;
export const SEND_VERIFICATION_CODE_URL = (itineraryId, phone, email, timestamp) =>
  `${DOMAIN}/hotels/itinerary/${itineraryId}/send-otp?mobile_number=${phone}&isPah=true&_=${timestamp}&email=${email}`;
export const OTP_VERIFICATION_URL = (itineraryId, otp, timestamp, email) =>
  `${DOMAIN}/hotels/itinerary/${itineraryId}/verify-otp?otp=${otp}&_=${timestamp}&email=${email}`;

// api call for getting cancellation policies of each room variant
export const HOTELS_ROOM_CANCELLATION_POLICY = `${DOMAIN}/hotels/cancelPolicy`;
