// ct-pwa/src/constants/api/trains.js
import { DOMAIN } from '../'
/* https://www.cleartrip.com/trains/auto_complete_station?station=ban */
export const AUTOCOMPLETE_STATIONS_URL = `${DOMAIN}/trains/auto_complete_station`

/* https://www.cleartrip.com/v2/api/trains/availability?
male_seniors=0&
number=12229&
end=NDLS&
distance=493&
date=2017/05/31&
start=LKO&
class=SL&
fare=178&
children=0&
female_seniors=1&
adults=0
*/
export const AVAILABILITY_URL = `${DOMAIN}/v2/api/trains/availability`

/* https://www.cleartrip.com/v2/api/trains/results
?male_seniors=0&
destination_station_code=NDLS&
date=1/05/2017&
nbs=1&
class=SL&
origin_station_code=LKO&
origin_city=Lucknow&
children=0&
female_seniors=0&
adults=1&
destination_city=New+Delhi
*/
export const SEARCH_URL = `${DOMAIN}/v2/api/trains/results`

export const CREATE_TRAIN_ITINERARY_URL = `${DOMAIN}/v2/api/trains/itinerary/create`

export const IRCTC_USER_STATUS_URL = `${DOMAIN}/trains/book/check_irctc_account_status`

export const TRAIN_TRAVELLERS_URL = itineraryId =>
  `${DOMAIN}/v2/api/trains/itinerary/${itineraryId}/travellers`

export const INITIATE_PAYMENT_URL = itineraryId =>
  `${DOMAIN}/v2/api/trains/itinerary/${itineraryId}/payment`

export const TRAIN_BOOKING_DETAILS_URL = itineraryId =>
  `${DOMAIN}/v2/api/trains/itinerary/${itineraryId}/appbookresponse`

export const TRAIN_IRCTC_INFO = `${DOMAIN}/trains/irctc/account/checkToken`
