// ct-pwa/src/constants/trains/paths.js
import { appRoutePrefix } from '../'

export const TRAINS_BASE_PATH = `${appRoutePrefix}/trains`

export const TRAINS_SEARCH = `${TRAINS_BASE_PATH}/search`

export const TRAINS_RESULTS = `${TRAINS_BASE_PATH}/results`

export const TRAINS_AVAILABILITY = `${TRAINS_BASE_PATH}/availability`

export const TRAINS_ITINERARY_BASE_PATH = itineraryId =>
  `${TRAINS_BASE_PATH}/itinerary/${itineraryId}`

export const TRAINS_ITINERARY_REVIEW = itineraryId =>
  `${TRAINS_ITINERARY_BASE_PATH(itineraryId)}/review`

export const TRAINS_ITINERARY_TRAVELLERS = itineraryId =>
  `${TRAINS_ITINERARY_BASE_PATH(itineraryId)}/travellers`

export const TRAINS_ITINERARY_TRAVELLERS_DETAILS = itineraryId =>
  `${TRAINS_ITINERARY_TRAVELLERS(itineraryId)}/details`

export const TRAINS_ITINERARY_TRAVELLERS_GST = itineraryId =>
  `${TRAINS_ITINERARY_TRAVELLERS(itineraryId)}/gst`

export const TRAINS_ITINERARY_PAYMENT = itineraryId =>
  `${TRAINS_ITINERARY_BASE_PATH(itineraryId)}/payment`

export const TRAINS_ITINERARY_PAYMENT_RETRY = itineraryId =>
  `${TRAINS_ITINERARY_BASE_PATH(itineraryId)}/payment_retry`

export const TRAINS_ITINERARY_IRCTC = itineraryId =>
  `${TRAINS_ITINERARY_BASE_PATH(itineraryId)}/irctc`

export const TRAINS_BOOKING_FAILED = itineraryId =>
  `${TRAINS_ITINERARY_BASE_PATH(itineraryId)}/booking_failed`

export const TRAINS_BOOKING_CONFIRMATION = itineraryId =>
  `${TRAINS_ITINERARY_BASE_PATH(itineraryId)}/confirmation`
